> èœğŸ£çš„ç¬¬ä¸€æ¬¡æ­£å¼ä¸Šæ‰‹Rï¼Œç›¸å½“äºç¿»è¯‘äº† `edgeR` è¿™ä¸ªåŒ…çš„ UserGuideï¼Œè·‘äº†ç¬¬ä¸€ä¸ª case
> çœ¼çœ‹å†™äº†è¿™ä¹ˆå¤šå­—ï¼Œè¿˜æ˜¯å‘ä¸€ä¸‹æ¯”è¾ƒå¥½ï¼Œä¸‡ä¸€æœ‰äººçœ‹å‘¢
![](https://upload-images.jianshu.io/upload_images/14383117-6023d7a38e434ae9.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


### edgeRç®€ä»‹

edgeR å¯ä»¥åº”ç”¨äºä»»ä½•å¯äº§ç”ŸåŸºå› ç»„ç‰¹å¾æ•°æ®ï¼ˆread countsï¼‰çš„æŠ€æœ¯ï¼Œèƒ½å¤Ÿä¸º RNA-seq å®éªŒä¸­è¯„ä¼°å·®å¼‚è¡¨è¾¾ã€ChIP-seq å®éªŒä¸­å·®å¼‚æ ‡è®°æä¾›ç»Ÿè®¡ç¨‹åºã€‚è¯¥RåŒ…å…·å¤‡é€‚ç”¨äºå¤šç»„å®éªŒçš„ç²¾å‡†ç»Ÿè®¡æ³•ï¼ŒåŒæ—¶è¿˜å…·å¤‡å¹¿ä¹‰çº¿æ€§æ¨¡å‹ï¼ˆglmsï¼‰çš„ç»Ÿè®¡å­¦æ–¹æ³•â€”â€”é€‚ç”¨äºä¸åŒå¤æ‚ç¨‹åº¦çš„å¤šå› ç´ å®éªŒã€‚æœ‰æ—¶äººä»¬å°†å‰è€…ç§°ä¸º classic edgeRï¼Œå°†åè€…ç§°ä¸º glm edgeRã€‚ç„¶è€Œä¸Šè¿°ä¸¤ç§æ–¹æ³•æ˜¯äº’è¡¥çš„ï¼Œå¹¶ä¸”æ—¶å¸¸åœ¨æ•°æ®åˆ†æä¸­è¢«ç»“åˆä½¿ç”¨ã€‚å¤§å¤šæ•° glm å‡½æ•°å¯ä»¥é€šè¿‡å‡½æ•°åç§°ä¸­çš„ "glm" è¯†åˆ«ï¼Œè¿™ç±»å‡½æ•°å¯åˆ©ç”¨ä¼¼ç„¶æ¯”æ£€éªŒæˆ–æ‹Ÿä¼¼ç„¶Fæ£€éªŒæ£€æµ‹å·®å¼‚è¡¨è¾¾ã€‚edgeR çš„åŠŸèƒ½çš„ä¸€ä¸ªé‡è¦ç‰¹ç‚¹æ˜¯ï¼Œä¸è®º classic å’Œ glm ä¸¤ç§æ–¹æ³•ï¼Œéƒ½å±äºç»éªŒè´å¶æ–¯æ–¹æ³•ï¼Œä»è€Œèƒ½å¤Ÿåœ¨å®éªŒåªå…·æœ‰æœ€å°æ°´å¹³çš„ç”Ÿç‰©å­¦é‡å¤æ—¶ï¼Œä¾ç„¶èƒ½å¤Ÿåˆ¤æ–­å‡ºåŸºå› ç‰¹å¼‚çš„ç”Ÿç‰©å­¦å·®å¼‚ã€‚edgeR å¯åº”ç”¨äºä¸åŒæ°´å¹³çš„å·®å¼‚è¡¨è¾¾ï¼Œå¦‚åŸºå› ã€å¤–æ˜¾å­ã€è½¬å½•æœ¬ã€æ ‡ç­¾ç­‰ï¼Œåˆ†æå¤–æ˜¾å­æ°´å¹³æ—¶å¯è½»æ˜“åœ°æ£€æµ‹å‡ºå¯å˜å‰ªåˆ‡æˆ–å¼‚æ„ä½“ç‰¹å¼‚æ€§å·®å¼‚è¡¨è¾¾ã€‚

### æ¡ˆä¾‹åˆ†æ

#### RNA-seq â€”â€” å£è…”è‚¿ç˜¤ vs å¯¹åº”çš„æ­£å¸¸ç»„ç»‡ (RNA-Seq of oral carcinomas vs matched normal)

åˆ†æçš„ç›®çš„æ˜¯æ£€æµ‹è‚¿ç˜¤å’Œæ­£å¸¸ç»„ç»‡å¯¹æ¯”ä¸‹å·®å¼‚è¡¨è¾¾çš„åŸºå› ï¼Œè¿™ä¸ªæ¡ˆä¾‹å¯ä»¥ä½“ç° edgeR ä¸­ GLM æ³•çš„å·¥ä½œèƒ½åŠ›ã€‚

##### 1.edgeR å®‰è£…

```R
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```

##### 2.æ•°æ®è¯»å–

ä¸‹è½½æ–‡ç«  [Tumor Transcriptome Sequencing Reveals Allelic Expression Imbalances Associated with Copy Number Alterations](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009317/) ä¸­çš„ Table S1ã€‚

- å»é™¤å¤šä½™åˆ—é¡¹ã€é‡å‘½ååˆ—å

  ```R
  > rawdata <- read.delim("TableS1.txt", check.names=FALSE, stringsAsFactors=FALSE)
  ```

  ```R
  > rawdata <- rawdata[,c(1:9)]
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-8a1246fce44b73d1.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/647/format/webp)

  ```R
  > names(rawdata)[1:9]<-c("RefSeqID","Symbol","NbrOfExons","8N","8T","33N","33T","51N","51T")
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-cf336c4aa5e63ecd.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/731/format/webp)

  > è¿™é‡Œä¸å°å¿ƒæŠŠè¡¨æ ¼çš„ä¸€äºŒè¡Œçš„è¡Œæ•°å¼„æ²¡äº†ï¼Œåé¢éœ€è¦æ³¨æ„æ•°å­—
  >
  > > ç»“æœåˆ°`DEGList`è¿™ä¸€æ­¥æ—¶å‡ºç°æŠ¥é”™â€œThe count matrix is a data.frame instead of a matrix and the first 6 columns are non-numeric.â€ ç”¨äº†as.matrix()åè¡¨æ ¼é‡Œæ‰€æœ‰å†…å®¹åˆéƒ½å˜æˆäº†factor...ğŸ˜“
  >
  > **æ‰€ä»¥ï¼Œæœ€åè¿˜æ˜¯ç›´æ¥åœ¨excelè¡¨æ ¼é‡Œè§£å†³äº†è¡ŒåğŸ’¨
> æ›´æ–°ï¼šå…¶å®è¿™ä¸ªå¾ˆå¥½è§£å†³ï¼Œç„¶è¿™æ˜¯ç¬¬ä¸€æ¬¡æ­£å¼ï¼Œæ²¡é”™â€œæ­£å¼â€ï¼Œç”¨R....



- åŠ è½½edgeR, ç”¨`DEGList`æ„å»ºåˆ—è¡¨

  ```R
  > library(edgeR) Loading 
  > y <- DGEList(counts=rawdata[,4:9], genes=rawdata[,1:3]) 
  ```

##### 3.æ³¨é‡Š

- ç­›é€‰è½¬å½•æœ¬

  è¿™ç¯‡æ–‡ç« å‘è¡¨å·²ç»æ˜¯å‡ å¹´å‰äº†ï¼Œæ•°æ®ä¸­ä¸€äº› RefSeq ID å¯èƒ½å’Œç°åœ¨ä¸€èˆ¬ä½¿ç”¨çš„ RefSeq ID æœ‰å‡ºå…¥ï¼Œæ‰€ä»¥åªéœ€è¦ä¿ç•™`org.HS.eg.db`ä¸­æœ‰çš„ã€å…·æœ‰ NCBI æ³¨é‡Šçš„é‚£éƒ¨åˆ†è½¬å½•æœ¬ã€‚

  ```R
  > library(org.Hs.eg.db) 
  > idfound <- y$genes$RefSeqID %in% mappedRkeys(org.Hs.egREFSEQ) 
  > y <- y[idfound,] 
  > dim(y) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-dfe0c811f47c36b6.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/544/format/webp)

  > å¯ä»¥çœ‹å‡ºæœ‰15548ä¸ªåŸºå› 

- å°† Entrez Gene ID åŠ å…¥æ³¨é‡Š

  ```R
  > egREFSEQ <- toTable(org.Hs.egREFSEQ) 
  > head(egREFSEQ) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-c258b98d77b30d92.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/504/format/webp)

  ```R
  > m <- match(y$genes$RefSeqID, egREFSEQ$accession) 
  > y$genes$EntrezGene <- egREFSEQ$gene_id[m] 
  ```

- åˆ©ç”¨ Entrez Gene ID æ›´æ–° gene symbol

  ```R
  > egSYMBOL <- toTable(org.Hs.egSYMBOL) 
  > head(egSYMBOL) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-41fbb2bc67412586.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/398/format/webp)

  ```R
  > m <- match(y$genes$EntrezGene, egSYMBOL$gene_id) 
  > y$genes$Symbol <- egSYMBOL$symbol[m] 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-f4fb9cbf12a0f950.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


##### 5.ç­›é€‰åŠå½’ä¸€åŒ– ï¼ˆNormalizationï¼‰

- ç­›é€‰å‡º count æ•°æœ€å¤šçš„è½¬å½•æœ¬

  æ¯ä¸ª gene symbol ä¿ç•™ä¸€ä¸ªè½¬å½•æœ¬

  ```R
  > o <- order(rowSums(y$counts), decreasing=TRUE) > y <- y[o,] 
  > d <- duplicated(y$genes$Symbol) 
  > y <- y[!d,] 
  > nrow(y) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-adcee651fd2bac4a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/401/format/webp) 

- é‡æ–°è®¡ç®—æ–‡åº“å¤§å°

  ```R
  > y$samples$lib.size <- colSums(y$counts) 
  ```

- å°† Use Entrez Gene ID è®¾ä¸ºè¡Œå

  ```R
  > rownames(y$counts) <- rownames(y$genes) <- y$genes$EntrezGene 
  > y$genes$EntrezGene <- NULL 
  ```

- TMMå½’ä¸€åŒ– (trimmed mean of M-values normalization)

  ```R
  > y <- calcNormFactors(y) 
  > y$samples 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-5aa94c60c1a94f1e.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/266/format/webp)

##### 6.æ•°æ®æŒ–æ˜ (Data Exploration)

- åˆ©ç”¨`plotMDS`ç»˜åˆ¶MDS (Multidimensional scaling) å›¾ 

  é¦–å…ˆåº”è¯¥åˆ†ææ ·æœ¬çš„ç¦»ç¾¤å€¼å’Œå…¶ä»–å…³ç³»ï¼Œå‡½æ•°`plotMDS`èƒ½å¤Ÿç”Ÿæˆæ ·æœ¬ä¹‹é—´è·ç¦»ä¸ç”Ÿç‰©å·®å¼‚ç³»æ•° (biological coefficient of variation, BCV) ç›¸å¯¹åº”çš„å›¾è¡¨ã€‚

  ```R
  > plotMDS(y) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-97ba942ac7b65a6d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/634/format/webp)å›¾ä¸­æ¨ªåæ ‡å°†è‚¿ç˜¤æ ·æœ¬å’Œæ­£å¸¸æ ·æœ¬åŒºåˆ†å¼€æ¥ï¼ˆå·¦Tå³Nï¼‰ï¼Œçºµåæ ‡åˆ™å¤§è‡´å¯¹åº”äº†æ‚£è€…ç¼–ç ï¼Œè¿™äº›åº”è¯äº†æ ·æœ¬çš„é…å¯¹ç‰¹æ€§ã€‚å¯ä»¥çœ‹å‡ºè‚¿ç˜¤æ ·æœ¬çš„åˆ†å¸ƒæ¯”æ­£å¸¸æ ·æœ¬æ›´ä¸å‡åŒ€ã€‚

##### 7.è®¾è®¡çŸ©é˜µ

- æ£€æµ‹å·®å¼‚è¡¨è¾¾

  åœ¨æ‹Ÿåˆè´ŸäºŒé¡¹ GLM ä¹‹å‰ï¼Œéœ€è¦æ ¹æ®å®éªŒè®¾è®¡ä¼˜åŒ–è®¾è®¡çŸ©é˜µã€‚æ¥ä¸‹æ¥éœ€è¦åœ¨åŒä¸€æ‚£è€…çš„è‚¿ç˜¤å’Œæ­£å¸¸ç»„ç»‡æ ·æœ¬ä¸­æ£€æµ‹å·®å¼‚è¡¨è¾¾ï¼Œå³è°ƒæ•´æ‚£è€…ä¹‹é—´çš„å·®å¼‚ã€‚ç»Ÿè®¡å­¦æ„ä¹‰ä¸Šï¼Œè¿™æ˜¯æŠŠæ‚£è€…ä½œä¸ºåŒºç»„å› å­çš„ä¸€ç§å¯åŠ çº¿æ€§æ¨¡å‹ã€‚

  ```R
  > Patient <- factor(c(8,8,33,33,51,51)) 
  > Tissue <- factor(c("N","T","N","T","N","T")) 
  > data.frame(Sample=colnames(y),Patient,Tissue)
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-c6b838f0ffa0d05c.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/415/format/webp)

  ```R
  > design <- model.matrix(~Patient+Tissue) 
  > rownames(design) <- colnames(y)
  > design
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-7da211b3778e535a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/411/format/webp)

  è¿™ç§å¯åŠ æ¨¡å‹é€‚ç”¨äºæˆå¯¹è®¾è®¡æˆ–å…·æ‰¹æ¬¡æ•ˆåº” (batch effect) çš„å®éªŒã€‚

##### 8.ä¼°è®¡æ•°æ®åˆ†å¸ƒ

- ä¼°è®¡æ•°æ®ç»„çš„è´ŸäºŒé¡¹åˆ†å¸ƒ

  ```R
  > y <- estimateDisp(y, design, robust=TRUE) 
  > y$common.dispersion 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-4eb757841b83deb8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/376/format/webp)

  å…¬å…±ç¦»æ•£å€¼ (common dispersion) çš„å¹³æ–¹æ ¹ï¼Œå³ç”Ÿç‰©å·®å¼‚  (biological variation) çš„å·®å¼‚ç³»æ•° (coefficient of variation)ï¼Œä¸º0.4ã€‚
  
  ```R
  > plotBCV(y) 
  ```
  
  ![](https://upload-images.jianshu.io/upload_images/14383117-4bbf3128e1264fd1.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/634/format/webp)

##### 9.å·®å¼‚è¡¨è¾¾

- æ‹ŸåˆåŸºå› å±‚é¢ GLM

  ```r
  > fit <- glmFit(y, design) 
  ```

- è‚¿ç˜¤æ ·æœ¬ vs æ­£å¸¸ç»„ç»‡ï¼Œåšä¼¼ç„¶æ¯”æ£€éªŒï¼Œåˆ—å‡ºå·®å¼‚æœ€ä¸ºæ˜¾è‘—çš„åŸºå› 

  ```r
  > lrt <- glmLRT(fit) 
  > topTags(lrt) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-1520a21b6cb63f24.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/676/format/webp)

  è¿™ä¸ªæ£€éªŒç±»ä¼¼é…å¯¹tæ£€éªŒã€‚æ’åˆ—é å‰çš„æ ‡ç­¾/åŸºå› çš„på€¼ã€FDRéƒ½å¾ˆå°ï¼Œå·®å¼‚å€æ•° (FC) å¾ˆå¤§ã€‚

  è¿›ä¸€æ­¥åœ°ğŸ‘‡

- å•ä¸ªæ ·æœ¬ä¸­è¿™äº›åŸºå› çš„æ¯ç™¾ä¸‡ count æ•° (counts-per-million)

  ps.æ ¹æ®på€¼æ’åº

  ```r
  > o <- order(lrt$table$PValue) 
  > cpm(y)[o[1:10],] 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-0c024aab77cd7ce1.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/610/format/webp)

  å¯ä»¥çœ‹å‡ºï¼Œè¿™äº›åŸºå› åœ¨è‚¿ç˜¤æ ·æœ¬å’Œæ­£å¸¸ç»„ç»‡ä¸­çš„å˜åŒ–ï¼Œåœ¨ä¸‰ä¸ªæ‚£è€…ä¸­æ˜¯ä¸€è‡´ã€‚

- ä»¥ FDRï¼ˆé”™è¯¯å‘ç°ç‡ï¼‰ä¸º5%ï¼ŒæŸ¥çœ‹å·®å¼‚è¡¨è¾¾åŸºå› çš„æ•°é‡

  ```r
  > summary(decideTests(lrt)) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-372436c57893356d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/233/format/webp)

  

- ç»˜åˆ¶ log-FC ä¸ log-cpm å‡å€¼ - å·®å¼‚å›¾ (mean-difference plot)
å›¾ä¸­çš„è“è‰²æ¨ªçº¿ä»£è¡¨2å€å·®å¼‚ (2-fold changes)ã€‚

  ```r
  > plotMD(lrt) 
  > abline(h=c(-1, 1), col="blue") 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-4eb24d1064fe923b.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


  

##### 10.GOåˆ†æ

- é’ˆå¯¹ç”Ÿç‰©å­¦è¿‡ç¨‹ (biological process, BP)åšGOåˆ†æ

  ```r
  > go <- goana(lrt) 
  > topGO(go, ont="BP", sort="Up", n=30, truncate=30) 
  ```

  ![](https://upload-images.jianshu.io/upload_images/14383117-890fa6b55b66c13e.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/700/format/webp)

  è‚¿ç˜¤ä¸­è¡¨è¾¾ä¸Šè°ƒçš„åŸºå› å¯èƒ½ä¸ç»†èƒåˆ†åŒ–ã€ç»†èƒè¿ç§»ã€ç»„ç»‡å½¢æ€å‘ç”Ÿç›¸å…³ã€‚

### Setup Info

![](https://upload-images.jianshu.io/upload_images/14383117-618cd7b24549d414.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/573/format/webp)

