> ![](https://upload-images.jianshu.io/upload_images/14383117-bcb81513d1a7a0a0.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

## affy: Methods for Affymetrix Oligonucleotide Arrays

> Gautier L, Cope L, Bolstad BM, Irizarry RA (2004). â€œaffyâ€”analysis of Affymetrix GeneChip data at the probe level.â€ *Bioinformatics*, **20**(3), 307â€“315. ISSN 1367-4803, doi: [10.1093/bioinformatics/btg405](https://doi.org/10.1093/bioinformatics/btg405).

## oligo: Preprocessing tools for oligonucleotide arrays

> Carvalho BS, Irizarry RA (2010). â€œA Framework for Oligonucleotide Microarray Preprocessing.â€ *Bioinformatics*, **26**(19), 2363-7. ISSN 1367-4803, doi: [10.1093/bioinformatics/btq431](https://doi.org/10.1093/bioinformatics/btq431).

## 1. å…³äºè¿™ä¸¤ä¸ªåŒ…

Jimmyå¤§ç¥æ˜¯è¿™ä¹ˆè¯´çš„ï¼š

> affyåŒ…æ˜¯Rè¯­è¨€çš„bioconductorç³»åˆ—åŒ…çš„ä¸€ä¸ªï¼Œå°±ä¸€ä¸ªåŠŸèƒ½ï¼Œè¯»å–affymetixçš„åŸºå› è¡¨è¾¾èŠ¯ç‰‡æ•°æ®-CELæ ¼å¼æ•°æ®ï¼Œå¤„ç†æˆè¡¨è¾¾çŸ©é˜µï¼ï¼ï¼
>
> oligoåŒ…æ˜¯Rè¯­è¨€çš„bioconductorç³»åˆ—åŒ…çš„ä¸€ä¸ªï¼Œå°±ä¸€ä¸ªåŠŸèƒ½ï¼Œè¯»å–affymetixçš„åŸºå› è¡¨è¾¾èŠ¯ç‰‡æ•°æ®-CELæ ¼å¼æ•°æ®ï¼Œå¤„ç†æˆè¡¨è¾¾çŸ©é˜µï¼ï¼ï¼

ä¸€äº›æœ¯è¯­ï¼š

> **probe** oligonucleotides of 25 base pair length used to probe RNA targets.
> **perfect match** probes intended to match perfectly the target sequence.
> ***PM*** intensity value read from the perfect matches.
> **mismatch** the probes having one base mismatch with the target sequence intended to account for non-specific binding.
> ***MM*** intensity value read from the mis-matches.
> **probe pair** a unit composed of a perfect match and its mismatch.
> **affyID** an identification for a probe set (which can be a gene or a fraction of a gene) represented on the array.
> **probe pair set** PMs and MMs related to a common affyID
>
> **CEL files** contain measured intensities and locations for an array that has been hybridized.
> **CDF file** contain the information relating probe pair sets to locations on the array.

é‚£ä¹ˆå°±æ¥**è¯»å–affymetixçš„åŸºå› è¡¨è¾¾èŠ¯ç‰‡æ•°æ®-CELæ ¼å¼æ•°æ®ï¼Œå¤„ç†æˆè¡¨è¾¾çŸ©é˜µï¼ï¼ï¼**

## 2. `affy`

### 2.1 è¯»å– *CEL* æ–‡ä»¶

æ–‡ä»¶æ¥è‡ª [GSE1428][ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE1nnn/GSE1428/suppl/GSE1428_RAW.tar]

ä¸¤ç§è¯»å–æ–¹æ³•ï¼š

- åœ¨ `celfile.path =` è®¾ç½® *CEL* æ–‡ä»¶æ‰€åœ¨è·¯å¾„

  ```R
  library(affy)
  dir_cels <- "YOUR PATH/affy/GSE1428_RAW"
  affyData <- ReadAffy(celfile.path = dir_cels)
  ```

- é€šè¿‡ `tkWidgets` åŒ…å¼¹å‡ºçª—å£

  ```R
  library(affy)
  library(tkWidgets)
  affyData <- ReadAffy(widget=TRUE)
  ```
![](https://upload-images.jianshu.io/upload_images/14383117-42b9033379d5593d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

  è¿™æ—¶å€™ç›´æ¥é¼ æ ‡ç‚¹å‡»é€‰æ‹©å³å¯...(*ï¿£ï¼ï¿£)ãƒ

### 2.2 Expression measures 

å°†æ¢é’ˆæ°´å¹³æ•°æ®è½¬æ¢ä¸ºè¡¨è¾¾é‡å€¼éœ€è¦ä»¥ä¸‹å‡ æ­¥ï¼š

> 1. reading in probe level data. 
>
> 2. background correction. 
>
> 3. normalization. 
>
> 4. probe speci c background correction, e.g. subtracting *MM*. 
>
> 5. summarizing the probe set values into one expression measure and, in some cases, a standard error for this summary. 

`affy` æœ‰ä¸€ç³»åˆ—å‡½æ•°å¯ä»¥å®Œæˆä»¥ä¸Šä»»åŠ¡ã€‚

#### 2.2.1 expresso

`expresso` çš„å‚æ•°é€‰æ‹©ï¼š

```r 
bgcorrect.methods()
# [1] "bg.correct" "mas"        "none"       "rma" 
normalize.methods(affyData)
# [1] "constant"         "contrasts"        "invariantset"    
# [4] "loess"            "methods"          "qspline"         
# [7] "quantiles"        "quantiles.robust"
pmcorrect.methods() 
# [1] "mas"        "methods"    "pmonly"     "subtractmm"
express.summary.stat.methods() 
# [1] "avgdiff"      "liwong"       "mas"          "medianpolish"
# [5] "playerout"   
```

åŒæ ·çš„ï¼Œä¹Ÿå¯ä»¥é€šè¿‡ `tkWidgets` åŒ…å¼¹å‡ºçª—å£è¿›è¡Œé€‰æ‹©ğŸ‘‡

```R
expresso(affyData, widget=TRUE) 
```

![](https://upload-images.jianshu.io/upload_images/14383117-4ab8f54d826044e8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

é€‰æ‹©å¥½å‚æ•°åï¼Œ`expresso` ä¸€é¡¿æ“ä½œå¾—åˆ°ä¸€ä¸ª `ExpressionSet` å¯¹è±¡ã€‚

```r
eset <- expresso(affyData, 
                 bgcorrect.method="rma",
                 normalize.method="qspline", 
                 pmcorrect.method="pmonly", 
                 summary.method="liwong")
# background correction: rma 
# normalization: qspline 
# PM/MM correction : pmonly 
#  expression values: liwong 
# background correcting...
# done.
# normalizing...[1] "samples= 496 k= 5 first= 999"
# ...
# done.
# 22283 ids to be processed
# |                    |
# |####################|
class(eset)
# [1] "ExpressionSet"
# attr(,"package")
# [1] "Biobase"
```

ç„¶åå°±å¾—åˆ°äº†**è¡¨è¾¾çŸ©é˜µ**ï¼ï¼ï¼

```R
expmtx_expso <- exprs(eset)
## ä¹Ÿå¯ä»¥é€‰æ‹©è¾“å‡ºä¸ºtxt
write.exprs(eset, file="mydata.txt") 
```

![](https://upload-images.jianshu.io/upload_images/14383117-245a3869290875a7.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

#### 2.2.2 MAS 5.0

å¦‚æœæƒ³ç”¨ MAS 5.0 ç®—æ³•ï¼Œä½¿ç”¨ `mas5()` å³å¯ã€‚

> **MAS 5.0** is the name of the Affymetrix algorithm used for producing gene expression signal (see [Affymetrix WebSite](http://www.affymetrix.com/)) for more details on the algorithm.
>
> The algorithm consists of background correction, calculation of the probe summary and scaling (typically termed probeset-level normalization). 

æ‰€ä»¥å°±æ˜¯è¿™æ ·çš„ï¼š

```R
eset_mas5 <- mas5(affyData)
# background correction: mas 
# PM/MM correction : mas 
# expression values: mas 
# background correcting...done.
# 22283 ids to be processed
# |                    |
# |####################|
```

#### 2.2.3 RMA

å¯¹åº”çš„ï¼ŒRMA ç®—æ³•æœ‰ `rma()` å‡½æ•°ã€‚

> **RMA** is an algorithm used to create an expression matrix from Affymetrix data. The raw intensity values are background corrected, log2 transformed and then quantile normalized. Next a linear model is fit to the normalized data to obtain an expression measure for each probe set on each array.

```R
eset_rma <- rma(affyData)
# Background correcting
# Normalizing
# Calculating Expression
```

### 2.3 æ•°æ®æŒ–æ˜ä¸­çš„è´¨é‡æ§åˆ¶

æŸ¥çœ‹ `ExpressionSet` å¯¹è±¡çš„ç›¸å…³ä¿¡æ¯ã€‚

```R
affyData
# AffyBatch object
# size of arrays=712x712 features (27 kb)
# cdf=HG-U133A (22283 affyids)
# number of samples=22
# number of genes=22283
# annotation=hgu133a
# notes=
phenoData(affyData)
# An object of class 'AnnotatedDataFrame'
#   sampleNames: GSM23735.CEL GSM23736.CEL ... GSM23757.CEL
#     (22 total)
#   varLabels: sample
#   varMetadata: labelDescription
pData(affyData)
# sample
# GSM23735.CEL      1
# GSM23736.CEL      2
# GSM23737.CEL      3
# GSM23738.CEL      4
# GSM23740.CEL      5
# GSM23741.CEL      6
# GSM23742.CEL      7
# GSM23743.CEL      8
# GSM23744.CEL      9
# GSM23745.CEL     10
# GSM23746.CEL     11
# GSM23747.CEL     12
# GSM23748.CEL     13
# GSM23749.CEL     14
# GSM23750.CEL     15
# GSM23751.CEL     16
# GSM23752.CEL     17
# GSM23753.CEL     18
# GSM23754.CEL     19
# GSM23755.CEL     20
# GSM23756.CEL     21
# GSM23757.CEL     22
```

å‡½æ•° `MAplot()` å¯å°†å¼ºåº¦æ•°æ®çš„æˆå¯¹å¯¹æ¯” (pairwise graphical) ä»¥å›¾åƒå½¢å¼å‘ˆç°å‡ºæ¥ã€‚

```
MAplot(affyData[1:4],pairs=TRUE,plot.method="smoothScatter")
```

![](https://upload-images.jianshu.io/upload_images/14383117-3c149e797a811ff3.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

#### 2.3.1 æŸ¥çœ‹ PM MM æ•°æ®

åˆ†åˆ«åˆ©ç”¨å‡½æ•° `pm()` åŠ `mm()` :

```R
affypm <- pm(affyData, gn[50])
```

![](https://upload-images.jianshu.io/upload_images/14383117-39d0d272a46b5a1c.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

```R
affymm <- mm(affyData, gn[50])
```

![](https://upload-images.jianshu.io/upload_images/14383117-11beaab905d49ed2.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

ä¸€ä¸ªå¿«é€Ÿå¾—åˆ° â€œå¤§äº PM çš„ MM çš„ç™¾åˆ†æ¯”â€ çš„æ–¹æ³•ï¼š

```R
mean(mm(affyData)>pm(affyData))
# [1] 0.3590052
```

## 3. `oligo`

### 3.1 è¯»å–æ•°æ®

`oligo` ä¸­ï¼Œ`list.celfiles()` ç”¨äºç”Ÿæˆ CEL æ–‡ä»¶åˆ—è¡¨ï¼Œ`read.celfiles()` ç”¨äºè¯»å– CEL æ–‡ä»¶ã€‚

```R
celFiles <- list.celfiles('YOUR PATH/GSE1428_RAW',
                          full.names=TRUE) 
rawData <- read.celfiles(celFiles) 
```

`oligo` å…·æœ‰å¤šç§ç±»å‹çš„ data container:

| Type                | Array      |
| ------------------- | ---------- |
| *ExonFeatureSet*    | Exon ST    |
| *ExpressionFeature* | Expression |
| *GeneFeatureSet*    | Gene ST    |
| *TilingFeatureSet*  | Tiling     |
| *SnpFeatureSet*     | SNP        |

### 3.2 å¤„ç†æˆè¡¨è¾¾çŸ©é˜µ

`rma()` åˆ©ç”¨ RMA ç®—æ³•è¿›è¡ŒèƒŒæ™¯æ ¡æ­£ã€å½’ä¸€åŒ–åŠ summarization. å¾—åˆ° `ExpressionSet` å¯¹è±¡ã€‚

```R
exset_rma <- rma(rawData)
# Background correcting
# Normalizing
# Calculating Expression
class(exset_rma)
# [1] "ExpressionSet"
# attr(,"package")
# [1] "Biobase"
```

ç„¶åå°±å¾—åˆ°äº†**è¡¨è¾¾çŸ©é˜µ**ï¼ï¼ï¼

```R
expmatx <- exprs(exset_rma)
```

![](https://upload-images.jianshu.io/upload_images/14383117-f7a740bf313a78d6.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

å½“ç„¶ä¹Ÿå¯ä»¥åˆ†æ­¥è¿›è¡Œï¼š

```r
backgroundCorrectionMethods() 
# [1] "rma"  "mas"  "LESN"
bgdata <- backgroundCorrect(rawData, method = 'mas' ) ## å®é™…ä¸Šé»˜è®¤å°±æ˜¯'mas'
normdata <- normalize(bgdata)
# Normalizing... OK
```

### 3.3 PM å¯è§†åŒ–

```R
boxplot(rawData, value = "pm")
```

![](https://upload-images.jianshu.io/upload_images/14383117-6e5d1bde90de9102.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

  ```R
boxplot(exset_rma, value = "pm")
  ```

![](https://upload-images.jianshu.io/upload_images/14383117-eaec37d42d662ec3.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

åŒæ ·çš„ï¼Œå‚æ•° `value = ` 'mm', 'bg', 'both', 'all' , è·å¾—å¯¹åº”çš„ boxplot.

### 3.4 æ¢é’ˆåºåˆ—

```R
pmseq <- pmSequence(rawData)
pmseq
#   A DNAStringSet instance of length 247965
#          width seq
#      [1]    25 TCGTTTTGGACTGACCTGGGGGTAT
#      [2]    25 CGAATTGTACGTGGGTCTGGACTTA
#      [3]    25 TGGACTAAACGAGATTTTAACCAGT
#      [4]    25 CCGACAGGGGTAGACGTTCCACTTG
#      [5]    25 AGGATTTAAGTCTGAGGTAGTCAAC
#      ...   ... ...
# [247961]    25 TTGTGAAGCACACAGCTCTGCAGCC
# [247962]    25 TTGGTGAGCACACTCCATTGACCAC
# [247963]    25 TTGGGAGAAGTGCCATTTCAGCAGA
# [247964]    25 TTGGGAAAGCATGGCACCCTCAGAC
# [247965]    25 TTGTGGACAGTGAGCACCGTCGTCA
```

å¾—åˆ° affinity splines coefficients, ç”» Sequence effect å›¾ï¼š

```R
pmsLog2 <- log2(pm(rawData)) 
coefs <- getAffinitySplineCoefficients(pmsLog2, pmseq) 
getBaseProfile(coefs[,1], plot = TRUE)
```

![](https://upload-images.jianshu.io/upload_images/14383117-33aa658e44e9a911.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

1, 2, 3, 4å¯èƒ½åˆ†åˆ«å¯¹åº” A, C, G, T.

ä¸ºä»€ä¹ˆè¯´å¯èƒ½å‘¢ï¼Œå› ä¸ºç¤ºä¾‹é‡Œçš„å›¾æ˜¯è¿™æ ·çš„...

![](https://upload-images.jianshu.io/upload_images/14383117-7e337c0cba7de2fa.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

åæ¥æ‰¾åˆ°äº†è€ç‰ˆçš„è¯´æ˜ä¹¦ï¼ŒåŸæ¥æ˜¯è¦å†åŠ äº›å‚æ•°...å®Œç¾è§£å†³ğŸ‘‡

```R
colors <- darkColors(4)
xL <- "Base Position"
yL <- "Effect"
pchs <- c("A", "C", "G", "T")
getBaseProfile(coefs[,1], plot = TRUE, 
               pch = pchs, type = "b", 
               xlab = xL, ylab = yL,
               lwd = 3, col = colors)
```

![](https://upload-images.jianshu.io/upload_images/14383117-6aead02df5369d55.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)



## References

- ç”¨affyåŒ…è¯»å–affymetixçš„åŸºå› è¡¨è¾¾èŠ¯ç‰‡æ•°æ®-CELæ ¼å¼æ•°æ® <http://www.bio-info-trainee.com/1580.html>

- ç”¨oligoåŒ…æ¥è¯»å–affymetixçš„åŸºå› è¡¨è¾¾èŠ¯ç‰‡æ•°æ®-CELæ ¼å¼æ•°æ® http://www.bio-info-trainee.com/1586.html

- MAS 5.0 - GENE User Manual 1.3 http://gene.moffitt.org/libaffy/doc/multi/MAS-5_002e0.html

- Robust Multi-array Average (RMA) http://www.molmine.com/magma/loading/rma.htm
- oligo User's Guide https://bioconductor.org/packages/release/bioc/vignettes/oligo/inst/doc/oug.pdf
- oligo - Primer https://www.bioconductor.org/packages//2.10/bioc/vignettes/oligo/inst/doc/primer.pdf

---
# æœ€åï¼Œå‘å¤§å®¶éš†é‡æ¨èç”Ÿä¿¡æŠ€èƒ½æ ‘çš„ä¸€ç³»åˆ—å¹²è´§ï¼

1.  ç”Ÿä¿¡æŠ€èƒ½æ ‘å…¨çƒå…¬ç›Šå·¡è®²ï¼š[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  Bç«™å…¬ç›Š74å°æ—¶ç”Ÿä¿¡å·¥ç¨‹å¸ˆæ•™å­¦è§†é¢‘åˆè¾‘ï¼š[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  æ‹›å­¦å¾’ï¼š[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)
