> å¤§ä½¬å°ä¸Šè®²è¯¾ä¸€åˆ†é’Ÿï¼Œå°ğŸŒ¶ğŸ¤è¯¾ä¸‹googleåå°æ—¶

> [Dragonstar2019 by Kai Wang](https://github.com/WGLab/dragonstar2019)
1.Detection of structural variants in human 
2.Annotation and phenotype-driven interpretation of genetic variants


## Detection of structural variants in human 

### Human genetic variation

![Pollex et al, Circulation. 2007](https://upload-images.jianshu.io/upload_images/14383117-e94ab58e51dbea0c.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

> ![image](https://upload-images.jianshu.io/upload_images/14383117-137f2d0738743621.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
>
> <http://doc.goldenhelix.com/SVS/tutorials/cnv_univariate_analysis/overview.html>

### Mechanisms underlying structural variant formation

> ![image](https://upload-images.jianshu.io/upload_images/14383117-83f4cae035735d44.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
>
> Ottaviani D, LeCain M, Sheer D. The role of microhomology in genomic structural variation[J]. Trends in Genetics, 2014, 30(3): 85-94.

#### Recurrent structural variants

- Share the same size and genomic content in unrelated individuals.

- Often caused by **NAHR** (Nonallelic homologous recombination--Nonallelic pairing of paralogous sequences and crossover leading to deletions, duplications and inversions )

  éç­‰ä½åŸºå› åŒæºé‡ç»„

  > **Non-allelic homologous recombination** (**NAHR**) is a form of [homologous recombination](https://en.wikipedia.org/wiki/Homologous_recombination) that occurs between two lengths of [DNA](https://en.wikipedia.org/wiki/DNA) that have high sequence similarity, but are not [alleles](https://en.wikipedia.org/wiki/Allele).
  >
  > <https://en.wikipedia.org/wiki/Non-allelic_homologous_recombination>

#### Nonrecurrent rearrangements

- Typical mechanisms:

  - NHEJ: Non-homologous end joining - éåŒæºæ€§æœ«ç«¯æ¥åˆ

    >**Nonhomologous end joining (NHEJ)** is an error-prone mechanism that does not use any DNA sequence homology to repair the DSB (Weterings & Chen, 2008).
    >
    >From: [Methods in Enzymology, 2017](https://www.sciencedirect.com/science/article/pii/S0076687917300927)

  - MMEJ: Microhomology-mediated end joining - å¾®åŒæºä»‹å°çš„æœ«ç«¯é€£æ¥

    > **Microhomology-mediated end joining** (**MMEJ**), also known as **alternative nonhomologous end-joining** (**Alt-NHEJ**) is one of the pathways for repairing [double-strand breaks](https://en.wikipedia.org/wiki/Double-strand_break) in DNA. 
    >
    > <https://en.wikipedia.org/wiki/Microhomology-mediated_end_joining>

  - FoSTeS/MMBIR: microhomology-mediated break-induced replication

    > FoSTeS: fork stalling and template switching

  - SRS: Smaller complex rearrangements caused by **serial replication slippage**

### Technologies for CNV Detection

#### Karyotyping and cytogenetic analysis

- Giemsa staining

  > ![image](https://upload-images.jianshu.io/upload_images/14383117-02f032ddbc48afd2.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

  - Giemsa banding (G-banding) - Gæ˜¾å¸¦
  - Dark bands are AT-rich and have less genes. 
  - Light bands are GC-rich DNA and are more transcriptionally active. 

- Fluorescent in situ hybridization (FISH) - è§å…‰åŸä½æ‚äº¤

  FISH uses **fluorescent probes** that bind to specific chromosomal 
  regions where there is a high degree of sequence complementarity.

- Comparative genomic hybridization (CGH) - æ¯”è¾ƒåŸºå› ç»„æ‚äº¤

  > **æ¯”è¾ƒåŸºå› ç»„æ‚äº¤**ï¼ˆè‹±è¯­ï¼šComparative genomic hybridizationï¼Œ**CGH**ï¼‰æ˜¯ä¸€ç§åˆ†å­[ç»†èƒé—ä¼ å­¦](https://zh.wikipedia.org/wiki/ç»†èƒé—ä¼ å­¦)æ–¹æ³•ï¼Œåœ¨ä¸åŸ¹å…»[ç»†èƒ](https://zh.wikipedia.org/wiki/ç»†èƒ)çš„æƒ…å†µä¸‹ï¼Œåˆ†æç›¸å¯¹äºå‚ç…§æ ·å“ï¼Œæµ‹è¯•æ ·å“çš„DNAä¸­[æ‹·è´æ•°å˜å¼‚](https://zh.wikipedia.org/w/index.php?title=æ‹·è´æ•°å˜å¼‚&action=edit&redlink=1)ï¼ˆCNVï¼‰çš„å¤šå€æ€§ç¨‹åº¦ã€‚å…¶ç›®çš„æ˜¯å¿«é€Ÿæœ‰æ•ˆåœ°æ¯”è¾ƒä¸¤ä¸ªæ¥æºçš„ä¸¤ç»„DNAæ ·æœ¬ï¼Œè¿™ä¸¤ç»„DNAé€šå¸¸æ˜¯å¯†åˆ‡ç›¸å…³çš„ï¼Œå› ä¸ºä¸¤è€…åœ¨æ•´ä¸ª[æŸ“è‰²ä½“](https://zh.wikipedia.org/wiki/æŸ“è‰²ä½“)æˆ–äºšæŸ“è‰²ä½“åŒºåŸŸï¼ˆæ•´ä¸ªæŸ“è‰²ä½“çš„ä¸€éƒ¨åˆ†ï¼‰ä¸Šéƒ½å¯èƒ½æœ‰è·å¾—æˆ–ä¸¢å¤±ã€‚è¯¥æŠ€æœ¯æœ€åˆæ˜¯ä¸ºäº†è¯„ä¼°å®ä½“ç˜¤å’Œæ­£å¸¸ç»„ç»‡çš„æŸ“è‰²ä½“äº’è¡¥å·®å¼‚è€Œå¼€å‘çš„[[1\]](https://zh.wikipedia.org/wiki/æ¯”è¾ƒåŸºå› ç»„æ‚äº¤#cite_note-Kallioniemi,Kallioniemi,Sudar,Rutovitz,Gray,Waldman,Pinkel-1)ï¼Œç›¸æ¯”æ›´ä¼ ç»Ÿçš„[Giemsaæ˜¾å¸¦](https://zh.wikipedia.org/wiki/Gæ˜¾å¸¦)æŠ€æœ¯å’Œ[è§å…‰åŸä½æ‚äº¤](https://zh.wikipedia.org/wiki/è§å…‰åŸä½æ‚äº¤)æŠ€æœ¯ï¼ˆFISHï¼‰ï¼ˆå—é™äºæ‰€ä½¿ç”¨çš„[æ˜¾å¾®é•œ](https://zh.wikipedia.org/wiki/æ˜¾å¾®é•œ)åˆ†è¾¨ç‡ï¼‰ï¼Œå®ƒçš„åˆ†è¾¨ç‡æé«˜åˆ°äº†5-10Mb[[1\]](https://zh.wikipedia.org/wiki/æ¯”è¾ƒåŸºå› ç»„æ‚äº¤#cite_note-Kallioniemi,Kallioniemi,Sudar,Rutovitz,Gray,Waldman,Pinkel-1)[[2\]](https://zh.wikipedia.org/wiki/æ¯”è¾ƒåŸºå› ç»„æ‚äº¤#cite_note-Strachan-2)[[3\]](https://zh.wikipedia.org/wiki/æ¯”è¾ƒåŸºå› ç»„æ‚äº¤#cite_note-Weiss,Hermsen,Meijer,VanGrieken,Baak,Kuipers,VanDiest-3)ã€‚
  >
  > https://zh.wikipedia.org/wiki/æ¯”è¾ƒåŸºå› ç»„æ‚äº¤

- Spectral karyotyping (SKY)

  > **Spectral karyotyping (SKY)** is a FISH-based method that labels each chromosome with a different color, allowing the identification of the chromosomal origin of all elements of the examined karyotype.
  >
  > From: [Molecular Diagnostics, 2010](https://www.sciencedirect.com/science/article/pii/B9780123694287000215)

### SNP genotyping arrays

SNP genotyping array is a type of **DNA microarray** which is used to **detect SNPs**. 
- Affymetrix arrays
  - In the Affymetrix assay, there are **25-mer probes** for both  alleles. 
  - Assuming there are two alleles (e.g. A-Allele and B-allele) at a 
    particular site. 
    - The DNA can bind to **both probes**.
    - But will have much **higher affinity** for the perfectly matched probe.
- Illumina arrays
  - In the Illumina array, attached to each Illumina **bead** is a **50**-
    **mer** sequence complementary to the sequence adjacent to the SNP site. 
  - The single-base extension (T or G) that is **complementary to the allele** carried by the DNA (A or C, respectively) then binds and results in the appropriately-colored signal (red or green, 
    respectively). 

> ![Schematic view of SNP array analysis by Affymetrix (right) and Illumina (left). ](https://upload-images.jianshu.io/upload_images/14383117-fc23cdbe7f4cb635.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
>
> Iacobucci I, Lonetti A, Papayannidis C, et al. Use of single nucleotide polymorphism array technology to improve the identification of chromosomal lesions in leukemia[J]. Current cancer drug targets, 2013, 13(7): 791-810.

### CNV Detection

#### Log R Ratio (LRR) and B Allele Frequency (BAF)

For both platforms, the computational algorithms convert the **raw signals** into **Log R Ratio (LRR)** and **B Allele Frequency (BAF)**. 

- **LRR** is a measure of normalized total signal intensity. 
- **BAF** is a measure of normalized allelic intensity ratio. 

> BAF = Y / (X + Y)
> LRR = log2( (X + Y)sampleOfInterest / (X+Y)baselineSample)
>
> <https://www.biostars.org/p/199025/>

#### Detection of CNVs from SNP arrays using *PennCNV*

- **Hidden Markov Model (HMM)** is a statistical Markov model in which the system being modeled is assumed to be a Markov process with hidden states.

  > **éšé©¬å°”å¯å¤«æ¨¡å‹**ï¼ˆHidden Markov Modelï¼›[ç¼©å†™](https://zh.wikipedia.org/wiki/ç¸®å¯«)ï¼š**HMM**ï¼‰æˆ–ç§°ä½œ**éšæ€§é©¬å°”å¯å¤«æ¨¡å‹**ï¼Œæ˜¯[ç»Ÿè®¡](https://zh.wikipedia.org/wiki/ç»Ÿè®¡)[æ¨¡å‹](https://zh.wikipedia.org/wiki/æ¨¡å‹)ï¼Œå®ƒç”¨æ¥æè¿°ä¸€ä¸ªå«æœ‰éšå«æœªçŸ¥å‚æ•°çš„[é©¬å°”å¯å¤«è¿‡ç¨‹](https://zh.wikipedia.org/wiki/é©¬å°”å¯å¤«è¿‡ç¨‹)ã€‚å…¶éš¾ç‚¹æ˜¯ä»å¯è§‚å¯Ÿçš„å‚æ•°ä¸­ç¡®å®šè¯¥è¿‡ç¨‹çš„éšå«å‚æ•°ã€‚ç„¶ååˆ©ç”¨è¿™äº›å‚æ•°æ¥ä½œè¿›ä¸€æ­¥çš„åˆ†æï¼Œä¾‹å¦‚[æ¨¡å¼è¯†åˆ«](https://zh.wikipedia.org/wiki/æ¨¡å¼è¯†åˆ«)ã€‚
  >
  > åœ¨**æ­£å¸¸çš„**é©¬å°”å¯å¤«æ¨¡å‹ä¸­ï¼ŒçŠ¶æ€å¯¹äºè§‚å¯Ÿè€…æ¥è¯´æ˜¯ç›´æ¥å¯è§çš„ã€‚è¿™æ ·çŠ¶æ€çš„è½¬æ¢æ¦‚ç‡ä¾¿æ˜¯å…¨éƒ¨çš„å‚æ•°ã€‚è€Œåœ¨**éš**é©¬å°”å¯å¤«æ¨¡å‹ä¸­ï¼ŒçŠ¶æ€å¹¶ä¸æ˜¯ç›´æ¥å¯è§çš„ï¼Œä½†å—çŠ¶æ€å½±å“çš„æŸäº›å˜é‡åˆ™æ˜¯å¯è§çš„ã€‚æ¯ä¸€ä¸ªçŠ¶æ€åœ¨å¯èƒ½è¾“å‡ºçš„ç¬¦å·ä¸Šéƒ½æœ‰ä¸€[æ¦‚ç‡åˆ†å¸ƒ](https://zh.wikipedia.org/wiki/æ¦‚ç‡åˆ†å¸ƒ)ã€‚å› æ­¤è¾“å‡ºç¬¦å·çš„åºåˆ—èƒ½å¤Ÿé€éœ²å‡ºçŠ¶æ€åºåˆ—çš„ä¸€äº›ä¿¡æ¯ã€‚
  >
  > https://zh.wikipedia.org/wiki/éšé©¬å°”å¯å¤«æ¨¡å‹
  >
  > å»¶ä¼¸é˜…è¯»ï¼š[æ¼«è°ˆ Hidden Markov Model](<http://freemind.pluskid.org/series/hmm-tutor/>)

- What we know are: LRR and BAF

- What we want to know is: **copy number**

Emission probability?

Transition probability?

LRR: æ€»çš„ä¿¡å·å¼ºåº¦

SNPâ†’CNVâ†’genotype call

SNP array æ•ˆç‡ä½

#### eg. Joint Modeling on Family Data

Joint modeling of the CNVs in a trio
### SV

#### NGS-based SV detection

Structural Variants

ç‰¹åˆ«å¤æ‚çš„SVéœ€è¦ de novo assembleï¼Œcontigæ¯”å¯¹åˆ°referenceä¸Š

#### Read count-based methods for SV detection
#### Detection of SVs from discordant read pairs
- Pattern of inversions
- tandem duplication
- translocations

#### Detection of SVs using assembly-based methods

**De novo sequence assembly (AS)** enables the fineï¿¾scale discovery of SVs, including novel (non-reference) sequence insertions

#### SV detection from long-read sequencing
#### Structural Variant Detection with Sniffles

second best alignment score is less than 2. å¯èƒ½æ˜¯100%ä¸€è‡´

ç»å¤§å¤šæ•°SVå¯ä»¥é€šè¿‡short read seqæ‰¾å‡ºæ¥

#### Bionano optical mapping for SV detection
##### SV detection from Single-molecule optical mapping 

## Annotation and phenotype-driven interpretation of genetic variants
### ANNOVAR for variant annotation

#### ANNOVAR supports 3 types of annotations:

- Gene-based annotation (g)

  Several commonly used **gene definitions**:

  - RefSeq gene - NCBI
  - UCSC known gene - UCSC
  - ENSEMBL gene - Europe
  - GENCODE gene - a project, all 

  **precedence rules**

- Region-based annotation (r)

  Several commonly used **filter annotation**:

  - Cytogenetic band - Keyword is cytoBand

  - Located in a ChIP-Seq peaks from ENCODE - å¾ˆå¤šå›½å®¶çš„ä¸€ä¸ªå®éªŒï¼Œé€šè¿‡ä¸åŒçš„ç»†èƒç³»???

  - Located in a predicted repressor, promoter, enhancer, etc

    chromHMM predictions é¢„æµ‹éç¼–ç åŒº?

    15 different â€œstatesâ€ are provided

    æŠŠåŸºå› ç»„åˆ†æˆäº†15ç§state

  - Overlap with conserved genomic regions

    - UCSC Genome Browser provides multi-species alignment for human 
      genome sequence
    - Conserved region may indicate functionally important regions

    *Selection of alignment*

- Filter-based annotation (f)

  Several commonly used filter annotation:

  - Allele frequency in various databases
  - Presence in various association databases
  - Functional prediction scores
  - dbSNP identifier

#### From functional annotation to clinical interpretation
-  5-tier system for clinical interpretation

- InterVar: Automated implementation of the ACMG-AMP clinical interpretation guidelines
### Introduction to HPO

The Human Phenotype Ontology (HPO) 

Our goal: from clinical phenotypes to genes

#### Phenolyzer

