# Dragon star Day 2 20190730
## å…³äºåºåˆ—æ¯”å¯¹ã€è®¡åˆ†æ–¹æ³•ã€BLASTåŸç†åŠåˆ¤æ–­å€¼ã€BWTåŸç†
>è¡¥ä¸ç¬”è®°ä¹‹ç»ˆäºä¸Šæ‰‹ç”»äº†
 -ä¸ºä»€ä¹ˆè¿™ä¹ˆéš¾ï¼Ÿ
-å› ä¸ºä½ æ˜¯æŒ‡å—é’ˆå‘€ 
-ã€‚
![](https://upload-images.jianshu.io/upload_images/14383117-8e30611f30521702.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


> [Dragonstar2019 by Kai Wang](https://links.jianshu.com/go?to=https%3A%2F%2Fgithub.com%2FWGLab%2Fdragonstar2019)
>
> 1. Alignment of short/long-read sequencing data
> 2. Genome assembly by short/long-read sequencing




## Part â…  Alignment of short/long-read sequencing data

### 1 Sequence similarity

>Sequence similarity is a measure of an empirical relationship between sequences.  
>
>A similarity score is therefore aimed to approximate the evolutionary distance between a pair of nucleotide or protein sequences.
>
>Heringa, Jaap(Jul 2008) Sequence Similarity. In: eLS. John Wiley & Sons Ltd, Chichester. http://www.els.net [doi: 10.1002/9780470015902.a0005317.pub2]

### 2 Sequence alignment

- Consider matches, mismatches, insertions and deletions 
  (indels are gaps)
- Find an **optimal** alignment between sequences
- Applications
  - Reference-based read mapping
  - Genome assembly
  - Gene finding

#### 2.1 Pairwise alignment

Sequence alignment between **two** sequences 

- Find an **optimal** alignment between two sequences

### 3 Dynamic programming

> **åŠ¨æ€è§„åˆ’**ï¼ˆè‹±è¯­ï¼šDynamic programmingï¼Œç®€ç§°DPï¼‰æ˜¯ä¸€ç§åœ¨[æ•°å­¦](https://zh.wikipedia.org/wiki/æ•°å­¦)ã€[ç®¡ç†ç§‘å­¦](https://zh.wikipedia.org/wiki/ç®¡ç†ç§‘å­¦)ã€[è®¡ç®—æœºç§‘å­¦](https://zh.wikipedia.org/wiki/è®¡ç®—æœºç§‘å­¦)ã€[ç»æµå­¦](https://zh.wikipedia.org/wiki/ç»æµå­¦)å’Œ[ç”Ÿç‰©ä¿¡æ¯å­¦](https://zh.wikipedia.org/wiki/ç”Ÿç‰©ä¿¡æ¯å­¦)ä¸­ä½¿ç”¨çš„ï¼Œé€šè¿‡æŠŠåŸé—®é¢˜åˆ†è§£ä¸ºç›¸å¯¹ç®€å•çš„å­é—®é¢˜çš„æ–¹å¼æ±‚è§£å¤æ‚é—®é¢˜çš„æ–¹æ³•ã€‚
>
> åŠ¨æ€è§„åˆ’å¸¸å¸¸é€‚ç”¨äºæœ‰é‡å å­é—®é¢˜å’Œ[æœ€ä¼˜å­ç»“æ„](https://zh.wikipedia.org/w/index.php?title=æœ€ä¼˜å­ç»“æ„&action=edit&redlink=1)æ€§è´¨çš„é—®é¢˜ï¼ŒåŠ¨æ€è§„åˆ’æ–¹æ³•æ‰€è€—æ—¶é—´å¾€å¾€è¿œå°‘äºæœ´ç´ è§£æ³•ã€‚
>
> åŠ¨æ€è§„åˆ’èƒŒåçš„åŸºæœ¬æ€æƒ³éå¸¸ç®€å•ã€‚å¤§è‡´ä¸Šï¼Œè‹¥è¦è§£ä¸€ä¸ªç»™å®šé—®é¢˜ï¼Œæˆ‘ä»¬éœ€è¦è§£å…¶ä¸åŒéƒ¨åˆ†ï¼ˆå³å­é—®é¢˜ï¼‰ï¼Œå†æ ¹æ®å­é—®é¢˜çš„è§£ä»¥å¾—å‡ºåŸé—®é¢˜çš„è§£ã€‚
>
> *https://zh.wikipedia.org/wiki/åŠ¨æ€è§„åˆ’*

- Purpose
  - To find an alignment between two sequences with best matching scores
- Three components
  - Recursive calculation - é€’å½’è®¡ç®—
  - Tabular arrangement - å¡«å……å¾—åˆ†çŸ©é˜µ
  - Traceback - è¿½æº¯æœ¬æº
- Three common types of pairwise alignments
    - Global alignment: Needleman-Wunsch
    - Local alignment: Smith-Waterman
    - Semi-global alignment

#### 3.1 Global alignment: Needleman-Wunsch

- **Best** global alignment

  Have **maximal alignment score** with all bases in two sequences

- Score function

  ![](https://upload-images.jianshu.io/upload_images/14383117-fc9a377ae5617202.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

##### 3.1.1 Working of Needleman -Wunsch Algorithm

> The dynamic programming matrix is defined with three different steps. 
>
> 1. Initialization of the matrix with the scores possible.
> 2. Matrix filling with maximum scores.
> 3. Trace back the residues for appropriate alignment.
>
> *vlab.amrita.edu,. (2012). Global alignment of two sequences - Needleman-Wunsch Algorithm. Retrieved 9 August 2019, from vlab.amrita.edu/?sub=3&brch=274&sim=1431&cnt=1* 

- Assume we have two sequences: P and Q

  - P = TCATGGC
  - Q = TCATC

- **Initial matrix**

  ![](https://upload-images.jianshu.io/upload_images/14383117-24eb084bde888621.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

- **Initialization Step**

  > Gap score is defined as penalty given to alignment, when we have insertion or deletion.
  >
  > This example assumes that there is gap penalty.  
  >
  > ä» Score function å¯çŸ¥ç©ºä½ç½šåˆ†ä¸º1
  >
  > *vlab.amrita.edu,. (2012). Global alignment of two sequences - Needleman-Wunsch Algorithm. Retrieved 9 August 2019, from vlab.amrita.edu/?sub=3&brch=274&sim=1431&cnt=1* 

  ![](https://upload-images.jianshu.io/upload_images/14383117-f9091520dcaf35bf.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

- **Matrix Fill Step**

  ![](https://upload-images.jianshu.io/upload_images/14383117-938ca4eacdaab249.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
  
  å…¨éƒ¨å†™å‡ºæ¥å¤ªä¹±äº†ï¼Œå°±å…ˆæ°æ ·å§
  
  ![](https://upload-images.jianshu.io/upload_images/14383117-7d2db908b9d2eb94.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

- **Trace back Step**

  > The final step in the algorithm is the trace back for the best alignment.
  >
  > By continuing the trace back step by the above defined method, one would reach to the 0th row, 0th column. 
  >
  >  The best alignment among the alignments can be identified by using the maximum alignment score which may be user defined.
  >
  > *vlab.amrita.edu,. (2012). Global alignment of two sequences - Needleman-Wunsch Algorithm. Retrieved 9 August 2019, from vlab.amrita.edu/?sub=3&brch=274&sim=1431&cnt=1* 

  ![](https://upload-images.jianshu.io/upload_images/14383117-fba904466d7ba332.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

  ![](https://upload-images.jianshu.io/upload_images/14383117-50eca70440a22d22.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

  å¤šç§æ¯”å¯¹ç»“æœï¼š

  ```
  purple line:
  TCATGGC
  ||||--|
  TCAT--C
  
  blue line:
  TCATGGC
  ||||-|-
  TCAT-C-
  
  yellow line:
  TCATGGC
  |||||--
  TCATC--
  ```


#### 3.2 Local alignment: Smith-Waterman

> **å²å¯†æ–¯-æ²ƒç‰¹æ›¼ç®—æ³•**ï¼ˆSmith-Waterman algorithmï¼‰æ˜¯ä¸€ç§è¿›è¡Œå±€éƒ¨åºåˆ—æ¯”å¯¹ï¼ˆç›¸å¯¹äºå…¨å±€æ¯”å¯¹ï¼‰çš„ç®—æ³•ï¼Œç”¨äºæ‰¾å‡ºä¸¤ä¸ªæ ¸è‹·é…¸åºåˆ—æˆ–è›‹ç™½è´¨åºåˆ—ä¹‹é—´çš„ç›¸ä¼¼åŒºåŸŸã€‚è¯¥ç®—æ³•çš„ç›®çš„ä¸æ˜¯è¿›è¡Œå…¨åºåˆ—çš„æ¯”å¯¹ï¼Œè€Œæ˜¯æ‰¾å‡ºä¸¤ä¸ªåºåˆ—ä¸­å…·æœ‰é«˜ç›¸ä¼¼åº¦çš„ç‰‡æ®µã€‚
>
> è¯¥ç®—æ³•ç”±[å¦æ™®å°”Â·å²å¯†æ–¯](https://zh.wikipedia.org/wiki/å¦æ™®å°”Â·å²å¯†æ–¯)ï¼ˆTemple F. Smithï¼‰å’Œ[è¿ˆå…‹å°”Â·æ²ƒç‰¹æ›¼](https://zh.wikipedia.org/wiki/è¿ˆå…‹å°”Â·æ²ƒç‰¹æ›¼)ï¼ˆMichael S. Watermanï¼‰äº1981å¹´æå‡ºã€‚å²å¯†æ–¯-æ²ƒç‰¹æ›¼ç®—æ³•æ˜¯[å°¼å¾·æ›¼-ç¿æ–½ç®—æ³•](https://zh.wikipedia.org/wiki/å°¼å¾·æ›¼-ç¿æ–½ç®—æ³•)çš„ä¸€ä¸ªå˜ä½“ï¼ŒäºŒè€…éƒ½æ˜¯åŠ¨æ€è§„åˆ’ç®—æ³•ã€‚è¿™ä¸€ç®—æ³•çš„ä¼˜åŠ¿åœ¨äºå¯ä»¥åœ¨ç»™å®šçš„æ‰“åˆ†æ–¹æ³•ä¸‹æ‰¾å‡ºä¸¤ä¸ªåºåˆ—çš„æœ€ä¼˜çš„å±€éƒ¨æ¯”å¯¹ï¼ˆæ‰“åˆ†æ–¹æ³•ä½¿ç”¨äº†ç½®æ¢çŸ©é˜µå’Œç©ºä½ç½šåˆ†ï¼‰ã€‚è¯¥ç®—æ³•å’Œå°¼å¾·æ›¼-ç¿æ–½ç®—æ³•çš„ä¸»è¦åŒºåˆ«åœ¨äºè¯¥ç®—æ³•ä¸å­˜åœ¨è´Ÿåˆ†ï¼ˆè´Ÿåˆ†è¢«æ›¿æ¢ä¸ºé›¶ï¼‰ï¼Œå› æ­¤å±€éƒ¨æ¯”å¯¹æˆä¸ºå¯èƒ½ã€‚å›æº¯ä»åˆ†æ•°æœ€é«˜çš„çŸ©é˜µå…ƒç´ å¼€å§‹ï¼Œç›´åˆ°é‡åˆ°åˆ†æ•°ä¸ºé›¶çš„å…ƒç´ åœæ­¢ã€‚åˆ†æ•°æœ€é«˜çš„å±€éƒ¨æ¯”å¯¹ç»“æœåœ¨æ­¤è¿‡ç¨‹ä¸­äº§ç”Ÿã€‚
>
> ![](https://upload-images.jianshu.io/upload_images/14383117-a781a0e3a5e9fd59.gif?imageMogr2/auto-orient/strip)
>
> https://zh.wikipedia.org/å²å¯†æ–¯-æ²ƒç‰¹æ›¼ç®—æ³•

- **Best** local alignment

  Have **maximal alignment score** with a subset of bases in two sequences

- Score function

  ![](https://upload-images.jianshu.io/upload_images/14383117-f8b1f4e205260d47.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

##### 3.2.1 Working of Smith-Waterman Algorithm

> The basic steps for the algorithm are similar to that of Needleman-Wunsch algorithm. The steps are:
>
> 1.	Initialization of a matrix.
> 2.	Matrix Filling with the appropriate scores.
> 3.	Trace back the sequences for a suitable alignment.
>
> *vlab.amrita.edu,. (2012). Smith-Waterman Algorithm - Local Alignment of Sequences. Retrieved 9 August 2019, from vlab.amrita.edu/?sub=3&brch=274&sim=1433&cnt=1*

å¯è§æ­¥éª¤éƒ½æ˜¯ä¸€è‡´çš„ï¼Œç”¨åŒæ ·çš„åºåˆ—å’Œ score function è¿›è¡Œæ¯”å¯¹
- Assume we have two sequences: P and Q

  - P = TCATGGC
  - Q = TCATC
  
- **Initial matrix**

  ![](https://upload-images.jianshu.io/upload_images/14383117-bbcce643e914811a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

- **Initialization Step**

  ![](https://upload-images.jianshu.io/upload_images/14383117-0a363fdbd6c8dcf7.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
  
- **Matrix Fill Step**

  ![](https://upload-images.jianshu.io/upload_images/14383117-f11387736d6c000a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

- **Trace back the sequences for a suitable alignment**

  1. Find a best score recursively
  2. Check which operation is used to obtain the current alignment score
  3. Stop when an alignment is 0

  æ‰€ä»¥æœ€ååº”è¯¥æ˜¯è¿™ä¸ªäºšå­ï¼š

  ![](https://upload-images.jianshu.io/upload_images/14383117-234d2870e454d8f2.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

  ```
  TCAT
  ||||
  TCAT
  ```


### 4 BLAST: The seed-index-map-extend-merge strategy

> The BLAST = **B**asic **L**ocal **A**lignment **S**earch **T**ool algorithm is a heuristic for computing optimal "local alignments" between a query sequence and a database containing one or more subject sequences.
>
> ![](https://upload-images.jianshu.io/upload_images/14383117-1074482487c11744.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
>
> <https://ab.inf.uni-tuebingen.de/teaching/ws2012/bioinf1/03-BLAST-BLAT.pdf>

#### 4.1 BLAST Basics

- BLAST is used for sequence **similarity** search, and much faster than Smith-Waterman method.
- Compare a query sequence against a **database** of sequences. 
- BLAST is a **collection** of algorithms
  - BLAST**N**: **nucleotide** sequence against **nucleotide** database
  - BLAST**P**: **protein** sequence against **protein** database
  - BLAST**X**: **translated** **nucleotide** sequence against **protein** database
  - **T**BLAST**X**: **six-frame translations of a nucleotide** query sequence 
    against the **six-frame translations of a nucleotide sequence** database 
  - **T**BLAST**N**: **protein** sequence against **translated nucleotide** database

#### 4.2 Process: The seed-index-map-extend-merge strategy

> Most popular algorithms use a seed-and-extend approach that operates in two steps:
> 1. Find a set of short exact matches called seeds.
> 2. Try to extend each seed match to obtain a long inexact match
>
> These matching words are called **seeds** and BLAST then attempts to extend each seed to a **HSP** (High-Scoring Segment Pair).
>
> https://ab.inf.uni-tuebingen.de/teaching/ws2012/bioinf1/03-BLAST-BLAT.pdf>

##### 4.2.1 The Process

- Seed and index

  - Construct common words (k-mers) for sequences in a database

  - Assume the database has two sequences and k=3

    > A â€œk-merâ€ is a word of DNA that is k long.
    >
    > Typically we extract k-mers from genomic assemblies or read data sets by running a k-length window across all of the reads and sequences.
    >
    > k-mers are most useful when theyâ€™re *long*, because then theyâ€™re *specific*. 
    >
    > The important concept here is that **long k-mers are species specific**. 
    >
    > <https://angus.readthedocs.io/en/2017/kmers-and-sourmash.html>

- Assume all 3-mers are high-score words

  BLAST needs a threshold to determine high-score words.

- Seed-map with high-score words

  Obtain words from a query sequence

- Extend-merge

  - Extend: high-scoring segment

  - BLAST output: 2 alignment containing match information and **E-value** for each alignment

    E-value çš„è§£é‡Šè§ 4.2.2

##### 4.2.2 Statistical significance of alignment

- How to assess the significance of a high-scoring hit to the database?

  > One crucial feature of the BLAST algorithm is that it gives an estimation of the statistical signficance of a determined **HSP**. This is based on so-called **Karlin-Altschul statistics** for local alignments (without gaps), which involves the **Poisson** distribution and other concepts.
  >
  > å³å­˜åœ¨ E-value, P-value
  >
  > https://ab.inf.uni-tuebingen.de/teaching/ws2012/bioinf1/03-BLAST-BLAT.pdf>

  **E** is the number of alignments expected by chance during a database search.

- Karlin-Altschul equation: $E(S) = Kmne^{-Î»S}$

  - Score (S) from scoring matrix
  - ğ¾ and Î» are two constants
  - ğ‘š: the length of a query sequence 
  - ğ‘›: the sum of bases of all sequences in database
  - The lower E value indicates more significant alignment.

  > E value represents the **number of matches with score â‰¥ S** that one would expect to see between two random sequences of lengths *n* and *m*.
  >
  > https://ab.inf.uni-tuebingen.de/teaching/ws2012/bioinf1/03-BLAST-BLAT.pdf>
  
- How to assess the P-value of finding a high-scoring pair (HSP)?

  - The number of random HSPs with score >= S is described by a **Poisson** distribution 

  - The **probability** of finding **exactly k HSPs** with score >=S is given by $P(X = k) = e^{âˆ’E} *E^k/k!$

  - The **probability** of finding **at least one HSP** by chance is 

    $P(S) = 1 âˆ’ P(X = 0) = 1 âˆ’ e^{âˆ’E}$

- E-value vs P-value
  - $E = âˆ’ ln(1 âˆ’ P)$
  - When E and P are very **small**, they are almost **identical**

### 5  BWT 

> Technically, it is a **lexicographical** reversible permutation of the characters of a string. 
>
> The most important application of BWT is found in biological sciences where genomes(long strings written in A, C, T, G alphabets) donâ€™t have many runs but they do have many repeats.
>
> <https://www.geeksforgeeks.org/burrows-wheeler-data-transform-algorithm/>

#### 5.1 Bowtie/BWA
> **Bowtie** is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).
>
> <http://bowtie-bio.sourceforge.net/index.shtml>

> **BWA** is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome. It consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM. 
>
> <http://bio-bwa.sourceforge.net/>

BWA and Bowtie are the most famous **implementations of BWT** in sequence alignment.

#### 5.2 BWT procedures

![](https://upload-images.jianshu.io/upload_images/14383117-6f38d6c35e308300.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

- Given a string S = â€œTCATCâ€, BWT
  - Adds `$` and assume `$`<any alphabet åœ¨æœ€ååŠ ä¸Š$
  - Obtains a suffix array of all cyclic rotations è½®è½¬åç¼€çŸ©é˜µ

- Cyclic rotation
  - Keeps an index of the rotated strings in the array å‰é¢çš„æ•°å­—
  - Creates Circular Permutation Table (CPT) è¡¨æ ¼æœ¬èº«
- Sorting
  - Sorts the Circular Permutation Table (CPT) **alphabetically** 
  - Keep the index with the strings æ•°å­—ä¸å˜ï¼Œæ¨ªå‘ä¸ºå•ä½
  - Taking the **last column** from the sorted array 
- BWT output
  - The last column represents the transformed string

#### 5.3 BWT reverse transformation

With a suffix array, it is easy to recover the **original string** S

![](https://upload-images.jianshu.io/upload_images/14383117-72ca4be97eff0577.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

---
# æœ€åï¼Œå‘å¤§å®¶éš†é‡æ¨èç”Ÿä¿¡æŠ€èƒ½æ ‘çš„ä¸€ç³»åˆ—å¹²è´§ï¼

1.  ç”Ÿä¿¡æŠ€èƒ½æ ‘å…¨çƒå…¬ç›Šå·¡è®²ï¼š[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  Bç«™å…¬ç›Š74å°æ—¶ç”Ÿä¿¡å·¥ç¨‹å¸ˆæ•™å­¦è§†é¢‘åˆè¾‘ï¼š[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  æ‹›å­¦å¾’ï¼š[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)








