> ğŸ˜‘
![](https://upload-images.jianshu.io/upload_images/14383117-39906d230e452e5a.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)



## å¯¼å…¥åŠæ£€ç´¢ Gene Model çš„ä¾¿æ·å·¥å…·

> Lawrence M, Huber W, PagÃ¨s H, Aboyoun P, Carlson M, Gentleman R, Morgan M, Carey V (2013). â€œSoftware for Computing and Annotating Genomic Ranges.â€ *PLoS Computational Biology*, **9**. doi: [10.1371/journal.pcbi.1003118](https://doi.org/10.1371/journal.pcbi.1003118), 

> *A **Gene Model** is defined as any description of a gene product from a variety of sources including computational prediction, mRNA sequencing, or genetic characterization.*

## 1. å…³äº `GenomicFeatures`

`GenomicFeatures` å¯ä»¥ä» UCSC Genome Bioinformatics åŠ BioMart ç­‰æ•°æ®åº“æ£€ç´¢å¹¶è·å–è½¬å½•æœ¬ç›¸å…³ä¿¡æ¯å¹¶å¯¹å…¶è¿›è¡Œæ“ä½œï¼Œæœ‰åŠ©äº ChIP-chip, ChIP-seq, RNA-seq ç»“æœåˆ†æã€‚

`GenomicFeatures` åˆ©ç”¨ `TxDb` å¯¹è±¡å‚¨å­˜è½¬å½•æœ¬å…ƒæ•°æ®ï¼ŒåŒ…æ‹¬ 5', 3' UTRs, CDS, å¤–æ˜¾å­ç­‰ã€‚åŒæ—¶ `TxDb` å¯¹è±¡æœ‰å¤§é‡è®¿é—®å‡½æ•°å¯è·å–è¿™äº›ä¿¡æ¯ã€‚

```R
library(GenomicFeatures)
ls("package:GenomicFeatures")
# [1] "as.list"                      "asBED"                       
# [3] "asGFF"                        "browseUCSCtrack"             
# [5] "cds"                          "cdsBy"                       
# [7] "cdsByOverlaps"                "coverageByTranscript"        
# [9] "DEFAULT_CIRC_SEQS"            "disjointExons"               
# [11] "distance"                     "exonicParts"                 
# [13] "exons"                        "exonsBy"                     
# [15] "exonsByOverlaps"              "extractTranscriptSeqs"       
# [17] "extractUpstreamSeqs"          "features"                    
# [19] "fiveUTRsByTranscript"         "genes"                       
# [21] "getChromInfoFromBiomart"      "getChromInfoFromUCSC"        
# [23] "getPromoterSeq"               "id2name"                     
# [25] "intronicParts"                "intronsByTranscript"         
# [27] "isActiveSeq"                  "isActiveSeq<-"               
# [29] "makeFDbPackageFromUCSC"       "makeFeatureDbFromUCSC"       
# [31] "makePackageName"              "makeTxDb"                    
# [33] "makeTxDbFromBiomart"          "makeTxDbFromEnsembl"         
# [35] "makeTxDbFromGFF"              "makeTxDbFromGRanges"         
# [37] "makeTxDbFromUCSC"             "makeTxDbPackage"             
# [39] "makeTxDbPackageFromBiomart"   "makeTxDbPackageFromUCSC"     
# [41] "mapFromTranscripts"           "mapIdsToRanges"              
# [43] "mapRangesToIds"               "mapToTranscripts"            
# [45] "microRNAs"                    "organism"                    
# [47] "pcoverageByTranscript"        "pmapFromTranscripts"         
# [49] "pmapToTranscripts"            "promoters"                   
# [51] "seqinfo"                      "seqlevels<-"                 
# [53] "seqlevels0"                   "show"                        
# [55] "species"                      "supportedMiRBaseBuildValues" 
# [57] "supportedUCSCFeatureDbTables" "supportedUCSCFeatureDbTracks"
# [59] "supportedUCSCtables"          "threeUTRsByTranscript"       
# [61] "tidyExons"                    "tidyIntrons"                 
# [63] "tidyTranscripts"              "transcriptLengths"           
# [65] "transcriptLocs2refLocs"       "transcripts"                 
# [67] "transcriptsBy"                "transcriptsByOverlaps"       
# [69] "transcriptWidths"             "tRNAs"                       
# [71] "UCSCFeatureDbTableSchema"    
```

## 2. ä» `TxDb` è·å–æ•°æ®

### 2.1 åŠ è½½è½¬å½•æœ¬æ•°æ®

- `loadDb()`
- ç›´æ¥åŠ è½½ TxDb æ³¨é‡ŠåŒ…

```R
samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite", 
                          package="GenomicFeatures") 
txdb_1 <- loadDb(samplefile) 
txdb_1 
# TxDb object:
# # Db type: TxDb
# # Supporting package: GenomicFeatures
# # Data source: UCSC
# # Genome: hg19
# # Organism: Homo sapiens
# # UCSC Table: knownGene
# # Resource URL: http://genome.ucsc.edu/
# # Type of Gene ID: Entrez Gene ID
# # Full dataset: no
# # miRBase build ID: NA
# # transcript_nrow: 178
# # exon_nrow: 620
# # cds_nrow: 523
# # Db created by: GenomicFeatures package from Bioconductor
# # Creation time: 2014-10-08 10:31:15 -0700 (Wed, 08 Oct 2014)
# # GenomicFeatures version at creation time: 1.17.21
# # RSQLite version at creation time: 0.11.4
# # DBSCHEMAVERSION: 1.0
```

```R
library(TxDb.Hsapiens.UCSC.hg19.knownGene) 
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene  
txdb 
# TxDb object:
# # Db type: TxDb
# # Supporting package: GenomicFeatures
# # Data source: UCSC
# # Genome: hg19
# # Organism: Homo sapiens
# # Taxonomy ID: 9606
# # UCSC Table: knownGene
# # Resource URL: http://genome.ucsc.edu/
# # Type of Gene ID: Entrez Gene ID
# # Full dataset: yes
# # miRBase build ID: GRCh37
# # transcript_nrow: 82960
# # exon_nrow: 289969
# # cds_nrow: 237533
# # Db created by: GenomicFeatures package from Bioconductor
# # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
# # GenomicFeatures version at creation time: 1.21.30
# # RSQLite version at creation time: 1.0.0
# # DBSCHEMAVERSION: 1.1
```
### 2.2  åŸºäºæŸ“è‰²ä½“çš„æ•°æ®é¢„ç­›é€‰

`seqlevels()` æŸ¥çœ‹æŸ“è‰²ä½“ï¼š

```R
head(seqlevels(txdb))
# [1] "chr1" "chr2" "chr3" "chr4" "chr5" "chr6"
```

é€šè¿‡èµ‹å€¼é€‰æ‹©æŸ“è‰²ä½“ï¼š

```R
seqlevels(txdb) <- "chr1" 
```

å¦‚æœéœ€è¦è®¾ç½®å›å…¨éƒ¨æŸ“è‰²ä½“çš„çŠ¶æ€ï¼Œä½¿ç”¨å‡½æ•° `seqlevels0()`ï¼š

```R
seqlevels(txdb) <- seqlevels0(txdb) 
```

### 2.3 æ•°æ®ç­›é€‰

`AnnotationDbi` åŒ…å†…çš„ `columns()` , `keytypes()` , `keys()` , `select()` åŒæ ·é€‚ç”¨äº `TxDb` å¯¹è±¡ã€‚

```R
## GENEID â†’ TXNAME
keys <- c("100033416", "100033417", "100033420") 
columns(txdb) 
#  [1] "CDSCHROM"   "CDSEND"     "CDSID"      "CDSNAME"   
#  [5] "CDSSTART"   "CDSSTRAND"  "EXONCHROM"  "EXONEND"   
#  [9] "EXONID"     "EXONNAME"   "EXONRANK"   "EXONSTART" 
# [13] "EXONSTRAND" "GENEID"     "TXCHROM"    "TXEND"     
# [17] "TXID"       "TXNAME"     "TXSTART"    "TXSTRAND"  
# [21] "TXTYPE"    
keytypes(txdb) 
# [1] "CDSID"    "CDSNAME"  "EXONID"   "EXONNAME" "GENEID"  
# [6] "TXID"     "TXNAME" 
select(txdb,
       keys = keys, 
       columns = "TXNAME",
       keytype = "GENEID") 
# 'select()' returned 1:1 mapping between keys and columns
#      GENEID     TXNAME
# 1 100033416 uc001yxl.4
# 2 100033417 uc001yxo.3
# 3 100033420 uc001yxr.3
```

```R
## GENEID â†’ TXNAME + TXTSTRAND + TXCHROM
cols <- c("TXNAME", "TXSTRAND", "TXCHROM") 
select(txdb, 
       keys = keys, 
       columns = cols, 
       keytype = "GENEID") 
# 'select()' returned 1:1 mapping between keys and columns
#      GENEID     TXNAME TXCHROM TXSTRAND
# 1 100033416 uc001yxl.4   chr15        +
# 2 100033417 uc001yxo.3   chr15        +
# 3 100033420 uc001yxr.3   chr15        +
```

## 3. ä¸»è¦å‡½æ•°åŠç¤ºä¾‹

> **Extract basic quantities**
>
> - `genes()`
> - `transcripts()`
> - `cds()`
> - `exons()`
> - `microRNAs()`
> - `tRNAs()`
> - `promoters()`
>
> **Extract quantities and group**
>
> - `transcriptsBy(by = c("gene", "exon", "cds"))`
> - `cdsBy(by = c("tx", "gene"))`
> - `exonsBy(by = c("tx", "gene"))`
> - `intronsByTranscript()`
> - `fiveUTRsByTranscript()`
> - `threeUTRsByTranscript()`
>
> (Note: there are grouping functions without the non-grouping function and vice versa; there is for example no `introns()` function.
>
> **Other functions**
>
> - `transcriptLengths()` (optionally include CDS length etc).
> - `XXByOverlaps()` (select features based on overlaps with `XX` being `transcript`, `cds` or `exon`).
>
> **Mapping between genome and transcript coordinates**
>
> - `extractTranscriptSeqs()` (getting RNA sequencing of the transcripts).

### 3.1 è¿”å› `GRanges` å¯¹è±¡

é™¤äº†**2.3** çš„æ•°æ®ç­›é€‰æ–¹æ³•ï¼Œç›´æ¥æå–å‡º `GRanges` å¯¹è±¡çš„æ–¹æ³•å¯èƒ½æ›´æ–¹ä¾¿ã€‚**"Extract basic quantities"** ä¸‹çš„æ‰€æœ‰å‡½æ•°éƒ½å¯è¿”å›å«æœ‰åŸºå› ç»„åæ ‡ã€å¤–æ˜¾å­åŒºé—´ã€è½¬å½•æœ¬ã€ç¼–ç åºåˆ—ç­‰ä¿¡æ¯çš„ `GRanges` å¯¹è±¡ã€‚

#### 3.1.1 `genes()`

```R
GE <- genes(txdb)
# [1] "GRanges"
# attr(,"package")
# [1] "GenomicRanges"
GE[1:5]
# GRanges object with 5 ranges and 1 metadata column:
#         seqnames              ranges strand |     gene_id
#            <Rle>           <IRanges>  <Rle> | <character>
#       1    chr19   58858172-58874214      - |           1
#      10     chr8   18248755-18258723      + |          10
#     100    chr20   43248163-43280376      - |         100
#    1000    chr18   25530930-25757445      - |        1000
#   10000     chr1 243651535-244006886      - |       10000
#   -------
#   seqinfo: 93 sequences (1 circular) from hg19 genome
```

#### 3.1.2 `transcripts()`

```R
TC <- transcripts(txdb) 
TC[1:3]
# GRanges object with 3 ranges and 2 metadata columns:
#       seqnames      ranges strand |     tx_id     tx_name
#          <Rle>   <IRanges>  <Rle> | <integer> <character>
#   [1]     chr1 11874-14409      + |         1  uc001aaa.3
#   [2]     chr1 11874-14409      + |         2  uc010nxq.1
#   [3]     chr1 11874-14409      + |         3  uc010nxr.1
#   -------
#   seqinfo: 93 sequences (1 circular) from hg19 genome
```

```R
## strand information
strand(TC)
# factor-Rle of length 82960 with 94 runs
#   Lengths: 4073 3894 2642 2450 2179 ...    2    1    2    2    1
#   Values :    +    -    +    -    + ...    -    +    -    +    -
# Levels(3): + - *
length(TC)
# 82960
```

ç”¨ `filter=` è¿›ä¸€æ­¥ç­›é€‰ï¼š

```R
TC_select <- transcripts(txdb,filter = list(tx_chrom = "chr15",
                                        tx_strand = "+"))
length(TC_select)
# [1] 1732
```

#### 3.1.3 `promoters()` 

```R
PR <- promoters(txdb, upstream = 2000, downstream = 400) 
PR[1:3]
# GRanges object with 3 ranges and 2 metadata columns:
#              seqnames     ranges strand |     tx_id     tx_name
#                 <Rle>  <IRanges>  <Rle> | <integer> <character>
#   uc001aaa.3     chr1 9874-12273      + |         1  uc001aaa.3
#   uc010nxq.1     chr1 9874-12273      + |         2  uc010nxq.1
#   uc010nxr.1     chr1 9874-12273      + |         3  uc010nxr.1
#   -------
#   seqinfo: 93 sequences (1 circular) from hg19 genome
```

#### 3.1.4 å¦ä¸€ç§æ–¹æ³•

```R
gr <- GRanges(seqnames = "chr1", 
              strand = "+", 
              ranges = IRanges(start = 11874, end = 14409))
subsetByOverlaps(exons(txdb), gr)
# GRanges object with 6 ranges and 1 metadata column:
#       seqnames      ranges strand |   exon_id
#          <Rle>   <IRanges>  <Rle> | <integer>
#   [1]     chr1 11874-12227      + |         1
#   [2]     chr1 12595-12721      + |         2
#   [3]     chr1 12613-12721      + |         3
#   [4]     chr1 12646-12697      + |         4
#   [5]     chr1 13221-14409      + |         5
#   [6]     chr1 13403-14409      + |         6
#   -------
#   seqinfo: 93 sequences (1 circular) from hg19 genome
```

å½“ç„¶ä¸­é—´çš„ `exons()` å¯ä»¥æ˜¯ä»»æ„ä¸€ä¸ª**"Extract basic quantities"** ä¸‹çš„å‡½æ•°ã€‚

### 3.2 ç‰¹å¾åˆ†ç»„

**Extract quantities and group** ä¸‹çš„å‡½æ•°å¯è¿”å› `GRangesList` å¯¹è±¡ï¼Œæ¯ä¸ªå‡½æ•°éƒ½æœ‰ç‰¹å®šçš„ `by=` å‚æ•°é€‰é¡¹ã€‚

```R
TC_bygene <- transcriptsBy(txdb, by = "gene") 
class(TC_bygene)
# [1] "CompressedGRangesList"
# attr(,"package")
# [1] "GenomicRanges"
length(TC_bygene)
# [1] 23459
names(TC_bygene)[10:13]
# [1] "10003"     "100033413" "100033414" "100033415"
```

```R
TC_bycds <- transcriptsBy(txdb, by = "cds")
length(TC_bycds)
# [1] 237533
names(TC_bycds)[10:13]
# [1] "10" "11" "12" "13"
TC_bycds[1:3]
# GRangesList object of length 3:
# $1 
# GRanges object with 1 range and 3 metadata columns:
#       seqnames      ranges strand |     tx_id     tx_name exon_rank
#          <Rle>   <IRanges>  <Rle> | <integer> <character> <integer>
#   [1]     chr1 11874-14409      + |         2  uc010nxq.1         1
# 
# $2 
# GRanges object with 1 range and 3 metadata columns:
#       seqnames      ranges strand | tx_id    tx_name exon_rank
#   [1]     chr1 11874-14409      + |     2 uc010nxq.1         2
# 
# $3 
# GRanges object with 1 range and 3 metadata columns:
#       seqnames      ranges strand | tx_id    tx_name exon_rank
#   [1]     chr1 11874-14409      + |     2 uc010nxq.1         3
# 
# -------
# seqinfo: 93 sequences (1 circular) from hg19 genome
```

åŒæ ·çš„ï¼Œå¦ä¸€ç§æ–¹æ³•ï¼š

```R
gr <- GRanges(seqnames = "chr1", 
              strand = "+", 
              ranges = IRanges(start = 11874, end = 14409))
subsetByOverlaps(cdsBy(txdb, by = "tx"), gr)
# GRangesList object of length 1:
# $2 
# GRanges object with 3 ranges and 3 metadata columns:
#       seqnames      ranges strand |    cds_id    cds_name exon_rank
#          <Rle>   <IRanges>  <Rle> | <integer> <character> <integer>
#   [1]     chr1 12190-12227      + |         1        <NA>         1
#   [2]     chr1 12595-12721      + |         2        <NA>         2
#   [3]     chr1 13403-13639      + |         3        <NA>         3
# 
# -------
# seqinfo: 93 sequences (1 circular) from hg19 genome
```

### 3.3 ä¸€æ­¥è·å¾—éç¼–ç åŒºåŠå†…å«å­çš„é•¿åº¦

`intronsByTranscript()` , `fiveUTRsByTranscript()` , `threeUTRsByTranscript()` æ˜¯åˆ†ç»„å‡½æ•°çš„â€œé¢„å…ˆè®¾å®šâ€ç‰ˆã€‚

```R
intronsByTranscript(txdb)
# GRangesList object of length 82960:
# $1 
# GRanges object with 2 ranges and 0 metadata columns:
#       seqnames      ranges strand
#          <Rle>   <IRanges>  <Rle>
#   [1]     chr1 12228-12612      +
#   [2]     chr1 12722-13220      +
# 
# $2 
# GRanges object with 2 ranges and 0 metadata columns:
#       seqnames      ranges strand
#   [1]     chr1 12228-12594      +
#   [2]     chr1 12722-13402      +
# 
# $3 
# GRanges object with 2 ranges and 0 metadata columns:
#       seqnames      ranges strand
#   [1]     chr1 12228-12645      +
#   [2]     chr1 12698-13220      +
# 
# ...
# <82957 more elements>
# -------
# seqinfo: 93 sequences (1 circular) from hg19 genome
```

```R
length(intronsByTranscript(txdb))
# [1] 82960
length(fiveUTRsByTranscript(txdb))
# [1] 61495
length(threeUTRsByTranscript(txdb))
# [1] 60740
```

## 4. è·å–åºåˆ—

å½“å’Œåˆé€‚çš„ `Bsgenome` åŒ…é…åˆä½¿ç”¨ï¼Œ`GenomicFeatures` ä¹Ÿå¯ä»¥ç›´æ¥è·å–åˆ°åºåˆ—æ•°æ®ã€‚

```R
library(BSgenome.Hsapiens.UCSC.hg19) 
```
è·å–è½¬å½•æœ¬åºåˆ—ï¼š

```R
tx_seqs1 <- extractTranscriptSeqs(Hsapiens, 
                                  TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                  use.names=TRUE) 
tx_seqs1
#   A DNAStringSet instance of length 82960
#          width seq                                  names               
#     [1]   1652 CTTGCCGTCAGCCTTTT...ACACTGTTGGTTTCTG uc001aaa.3
#     [2]   1488 CTTGCCGTCAGCCTTTT...ACACTGTTGGTTTCTG uc010nxq.1
#     [3]   1595 CTTGCCGTCAGCCTTTT...ACACTGTTGGTTTCTG uc010nxr.1
#     [4]    918 ATGGTGACTGAATTCAT...TAGTGTAAAGTTTTAG uc001aal.1
#     [5]     32 TACAGACCAAGCTCATGACTCACAATGGCCTA     uc001aaq.2
#     ...    ... ...
# [82956]   1217 GCCAGTTTAGGGTCTCT...TCCGCCTCCTGAGATC uc011mgu.1
# [82957]    737 CTCCACTTCTGATCCTC...TTTATTAGATGCAGTG uc011mgv.2
# [82958]     30 TGGTGAATTTCAGCCAAAGTGGCCAAAGAA       uc011mgw.1
# [82959]     30 TTGCAGAGGTGGCTGGTTGCTCTTTGAGCC       uc022brq.1
# [82960]     30 TGGTGAATTTCAGCCAAAGTGGCCAAAGAA       uc022brr.1

```

è¿˜å¯ä»¥ç¿»è¯‘ï¼š

```R
translate(tx_seqs1)
#  A AAStringSet instance of length 82960
#         width seq                                   names               
#     [1]   550 LAVSLFFDLFFLFMCIC...LHPAQLEILY*KHTVGF uc001aaa.3
#     [2]   496 LAVSLFFDLFFLFMCIC...ASCTARDPLLKAHCWFL uc010nxq.1
#     [3]   531 LAVSLFFDLFFLFMCIC...LHPAQLEILY*KHTVGF uc010nxr.1
#     [4]   306 MVTEFIFLGLSDSQELQ...AIRQLRKWDAHSSVKF* uc001aal.1
#     [5]    10 YRPSS*LTMA                            uc001aaq.2
#     ...   ... ...
# [82956]   405 ASLGSLVSPAELLCSRL...LRRASMATDHCNLRLLR uc011mgu.1
# [82957]   245 LHF*SSPPWCCSGSHTR...KWENGFLGLK*PLY*MQ uc011mgv.2
# [82958]    10 W*ISAKVAKE                            uc011mgw.1
# [82959]    10 MQRWLVAL*A                            uc022brq.1
# [82960]    10 W*ISAKVAKE                            uc022brr.1
```
























## References

- **Making and Utilizing TxDb Objects** https://bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf

- **GenomicFeatures Man Pages** <https://kasperdanielhansen.github.io/genbioconductor/html/GenomicFeatures.html#examples>

- **Using the Gene Search Results** <https://www.arabidopsis.org/help/helppages/generesu.jsp>
---
# æœ€åï¼Œå‘å¤§å®¶éš†é‡æ¨èç”Ÿä¿¡æŠ€èƒ½æ ‘çš„ä¸€ç³»åˆ—å¹²è´§ï¼

1.  ç”Ÿä¿¡æŠ€èƒ½æ ‘å…¨çƒå…¬ç›Šå·¡è®²ï¼š[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  Bç«™å…¬ç›Š74å°æ—¶ç”Ÿä¿¡å·¥ç¨‹å¸ˆæ•™å­¦è§†é¢‘åˆè¾‘ï¼š[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  æ‹›å­¦å¾’ï¼š[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)
