> å®åœ¨æ‡’å¾—æ‰¾memeäº†
![](https://upload-images.jianshu.io/upload_images/14383117-f523fc5dd04cf591.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


# GOrilla, clusterProfiler, topGO åˆ†æ ranked gene list (GSEA) å¯¹æ¯” 

## Genelist preparation

ä¿è¯äº†ä¸‰ä¸ªå·¥å…·è¾“å…¥çš„åŸºå› æ˜¯ä¸€è‡´çš„ã€‚

```r
# genelist preparation
library(hgu133plus2.db)
rankedprobe <- row.names(DEGdf)
load("GSE76275_DEGs.Rdata")
## ranked entrezid by pval
rankedentrezt <- AnnotationDbi::select(hgu133plus2.db,
                                      rankedprobe,
                                      "ENTREZID",
                                      "PROBEID")
rankedentrez <- unique(na.omit(rankedentrezt$ENTREZID))
```

### for *GOrilla*

çº æ­£ä¸€ä¸ªè¯¯è§£ï¼Œä¹‹å‰ä»¥ä¸º *GOrilla* è¦æ±‚çš„ gene list æ˜¯æŒ‰è¡¨è¾¾é‡ä»é«˜åˆ°ä½æ’åºã€‚

> This is particularly useful in many typical cases where genomic data may be naturally represented as a ranked list of genes (e.g. by level of expression or of differential expression). 

ä½†å®é™…ä¸Šæ˜¯ä¾ç„¶æ˜¯åŸºäºtæ£€éªŒï¼Œä¹Ÿå°±æ˜¯åŸºäºpå€¼ã€‚ 

> All genes were ranked according to how well they differentiate between the two groups using a simple t-test. The top of the list contained the genes that were the best separators between the two groups. 

*GOrilla* çš„ä¼˜åŠ¿ä¸»è¦ä½“ç°åœ¨ï¼š

- ä¸åŸºäºè¶…å‡ ä½•åˆ†å¸ƒçš„å·¥å…·ç›¸æ¯”â€”â€”é˜ˆå€¼çµæ´»
- ä¸å…¶ä»– é˜ˆå€¼çµæ´»/ç±»ä¼¼GSEA/åˆ©ç”¨Kolmogorov-Smirnovæ£€éªŒ çš„å·¥å…·ç›¸æ¯”ï¼š
  - ä¸“æ³¨äºGOï¼Œå¯ç›´æ¥ç”ŸæˆDAG
  - åˆ©ç”¨ "mHG " å¯å¾—åˆ°ç²¾ç¡®çš„ p.val
  - é€Ÿåº¦å¿«
  - ï¼ˆä¼¼ä¹æ”¾åœ¨ç°åœ¨è¿™äº›éƒ½ä¸æ˜¯ä»€ä¹ˆé—®é¢˜ï¼Œæ¯•ç«Ÿ GSEA 2005å¹´å‘å¸ƒï¼Œ*GOrilla* 2008å¹´å‘å¸ƒ...ä¸è¿‡ä¸€å¸ƒæ‹¿åˆ°å›¾+è¡¨ä¹Ÿæ˜¯å¾ˆçˆ½çš„é¢ (ã€‚ï¹ã€‚)

```R
## ranked symbol (increasing by pval)
rankedsymbol <- AnnotationDbi::select(hgu133plus2.db,
                                       rankedentrez,
                                       "SYMBOL",
                                       "ENTREZID")[,2]
save(rankedsymbol, file = "rankedsymbol_pvbased_forGOrilla.Rdata")
write.table(rankedsymbol, file = 
              "rankedsymbol_pvbased_forGOrilla.txt",
            quote = F, col.names = F, row.names = F)
```

æ‰‹åŠ¨é€‰æ‹© pvalue cutoff

![](https://upload-images.jianshu.io/upload_images/14383117-9ef4602d458485b4.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

### for clusterProfiler

```R
## ranked entrezid with pval (increasing by pval)
matchedprobe <- match(rankedentrez,
                      rankedentrezt$ENTREZID)
matchedprobe2 <- rankedentrezt[matchedprobe,][,"PROBEID"]
respon.pval <- DEGdf[matchedprobe2,]$P.Value
entrezlist_pval <- respon.pval
names(entrezlist_pval) <- rankedentrez
save(entrezlist_pval, file = "genelist_ENTREZID_pVal.Rdata")

### decreasing entrezid list by pval
entrezlist_dcrpv <- sort(entrezlist_pval, decreasing = T)
save(entrezlist_dcrpv, file = "genelist_ENTREZID_Decr_pVal.Rdata")

### decreasing entrezid list by FC
entrezlist_FC <- DEGdf[matchedprobe2,]$logFC
names(entrezlist_FC) <- rankedentrez
entrezlist_FC <- sort(entrezlist_FC, decreasing = T)
save(entrezlist_FC, file = "genelist_ENTREZID_FC.Rdata")
```

### for topGO 

```R
## probeid with p.val list (increasing by pval)
probelist_pval <- respon.pval
names(probelist_pval) <- matchedprobe2
save(probelist_pval, file = "genelist_PROBEID_pVal.Rdata")
```

## *GOrilla* 

### Running mode: Single ranked list of genes

å¾ˆé…·ç‚«çš„ä¸€å¼ **å¤§**å›¾

![](https://upload-images.jianshu.io/upload_images/14383117-e6936e8cf569c028.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![](https://upload-images.jianshu.io/upload_images/14383117-916cb91302d3e556.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

é‡å¤åŸºå› ä¼šè¢«è‡ªåŠ¨åˆ é™¤

![](https://upload-images.jianshu.io/upload_images/14383117-222a2dddddf4ec3a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

ä½†å¹¶ä¸çŸ¥é“å“ªæ¥çš„è¿™12ä¸ª duplicate genes (+_+)?

```R
table(table(rankedsymbol))
# 
#     1 
# 22012 
```

> **Enrichment (N, B, n, b)** is defined as follows:
> N - is the total number of genes
> B - is the total number of genes associated with a specific GO term
> n - is the number of genes in the top of the user's input list or in the target set when appropriate
> b - is the number of genes in the intersection
> Enrichment = (b/n) / (B/N)

![](https://upload-images.jianshu.io/upload_images/14383117-ff5c49256816585e.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

ç¼ºç‚¹æ˜¯å¹¶æ²¡æœ‰æ˜¾ç¤ºå¯Œé›†åˆ°äº†å¤šå°‘ä¸ªGO term, å¼ºè¡Œæ‰’ä¸‹è¿™ä¸ªè¡¨åï¼Œå°å°æ“ä½œï¼š

```R
GOrillares <- read.csv("GOrillatable.txt", header = T, sep = "\t")
save(GOrillares, file = "GOrillatable.Rdata")
```

![](https://upload-images.jianshu.io/upload_images/14383117-a0f52899c3a7bda5.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

æ‰€ä»¥å°±æ˜¯æœ‰100ä¸ª term.

```R
nrow(GOrillares)
# [1] 100
```

## clusterProfiler 

### GSEA GO-BP

æ ‡å‡†çš„ GSEA.

å…³äºè¾“å…¥çš„ gene list, yå”æ˜¯è¿™ä¹ˆè¯´çš„ï¼š

> The `geneList` contains three features:
>
> 1. numeric vector: fold change or other type of numerical variable
> 2. named vector: every number was named by the corresponding gene ID
> 3. sorted vector: number should be sorted in decreasing order

ä½†ç”¨ p.val çš„æ•ˆæœä¼¼ä¹ä¸€èˆ¬ã€‚

```R
## clusterProfiler GSEA GO-BP
library(clusterProfiler)
library(hgu133plus2.db)
load("genelist_ENTREZID_Decr_pVal.Rdata")

### with decreasing p.val
gseaGO_BP <- gseGO(geneList     = entrezlist_dcrpv,
                   OrgDb        = hgu133plus2.db,
                   keyType      = "ENTREZID",
                   ont          = "BP",
                   nPerm        = 1000,   ## æ’åˆ—æ•°
                   minGSSize    = 5,
                   maxGSSize    = 500,
                   pvalueCutoff = 0.95,
                   verbose      = TRUE)  ## ä¸è¾“å‡ºç»“æœ
gseaGO_BPresult <- gseaGO_BP@result
save(gseaGO_BPresult, file = "cp_gseaGO_BPresult.Rdata")
```

å¯èƒ½å’Œç½‘ç»œä¹Ÿæœ‰å…³ç³»ï¼Ÿå‡ æ¬¡è¾“å…¥åŒæ ·çš„å‚æ•°ï¼Œå¾—åˆ°çš„ç»“æœå¹¶ä¸ä¸€æ ·....æœ‰æ—¶æ˜¯æŠ¥é”™ no term, è¿·æƒ‘desu

![](https://upload-images.jianshu.io/upload_images/14383117-9567e31918bf40f8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
![](https://upload-images.jianshu.io/upload_images/14383117-22c9461a2422d778.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

ä¹Ÿæ²¡ä»€ä¹ˆæ˜¾è‘—çš„ term

![](https://upload-images.jianshu.io/upload_images/14383117-c6d396cb66815f82.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

è¿˜æ˜¯åƒç¤ºä¾‹é‡Œé‚£æ ·ç”¨ decreasing FC 

```r
### with decreasing FC
load("genelist_ENTREZID_FC.Rdata")
gseaGO_BP2 <- gseGO(geneList     = entrezlist_FC,
                    OrgDb        = hgu133plus2.db,
                    keyType      = "ENTREZID",
                    ont          = "BP",
                    nPerm        = 1000,   ## æ’åˆ—æ•°
                    minGSSize    = 5,
                    maxGSSize    = 500,
                    pvalueCutoff = 0.05,
                    verbose      = TRUE)  ## ä¸è¾“å‡ºç»“æœ
gseaGO_BPresult2 <- gseaGO_BP2@result
save(gseaGO_BPresult2, file = "cp_gseaGO_BPresult2.Rdata")
```

ç»“æœæœç„¶æ­£å¸¸å¤šäº†

![](https://upload-images.jianshu.io/upload_images/14383117-8e6bf3b718055265.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

## topGO 

### GSEA-like GO-BP (Using the genes score )

å‡†ç¡®åœ°è¯´ï¼Œä¹Ÿè®¸ä¸èƒ½ç›´æ¥å«åš GSEA æ³•ã€‚topGO è¯´æ˜ä¹¦é‡Œçš„è¯´æ³•æ˜¯ï¼š

> We will use two types of test statistics: Fisher's exact test which is based on gene counts, and a Kolmogorov-Smirnov like test which computes enrichment based on gene scores. 

å¹¶ä¸”ï¼š

> We can use both these tests since each gene has a score (representing how di erentially expressed a gene is) and by the means of topDiffGenes functions the genes are categorized into di erentially expressed or not di erentially expressed genes.

æ­£å› ä¸ºæœ‰ `topDiffGenes` è¿™ä¸ªå‡½æ•°çš„å­˜åœ¨ï¼Œç›¸å½“äºæ— å½¢ä¸­ç»™äº† target list, è€Œ `allGenes = `  ç›¸å½“äº universe gene list. æ‰€ä»¥è¿™ä¹Ÿè®¸æ˜¯ `topGO` æ— è®ºç”¨ç±»ä¼¼ ORA æˆ– ç±»ä¼¼ GSEA æ³•æ„å»ºçš„ `topGOdata` ï¼Œç®—æ³•å’Œæ£€éªŒæ–¹æ³•å¯ä»¥å„ç§æ··ç”¨çš„åŸå› ã€‚(maybe <@_@>

![](https://upload-images.jianshu.io/upload_images/14383117-ab474b6f0e2e5e11.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

```R
## topGO GSEA GO-BP
library(topGO)
library(hgu133plus2.db)
load("genelist_PROBEID_pVal.Rdata")

## function: topDiffGenes
topDiffGenes <- function(allScore) {
  return(allScore < 0.01)   ## p.val < 0.01
}
sum(topDiffGenes(probelist_pval))
# [1] 12210

## construct a topGOdata object
GSEAGOdata_BP <- new("topGOdata",
                     ontology = "BP",
                     allGenes = probelist_pval,
                     geneSel = topDiffGenes,
                     nodeSize = 5,
                     annot = annFUN.db,
                     affyLib = "hgu133plus2.db")
```

```R
## Kolmogorov-Smirnov like test - classic
KSres <- runTest(GSEAGOdata_BP, algorithm = "weight01", statistic = "ks") 
KSres
# 
# Description:  
# Ontology: BP 
# 'weight01' algorithm with the 'ks' test
# 9106 GO terms scored: 203 terms with p < 0.01
# Annotation data:
#     Annotated genes: 16454 
#     Significant genes: 9542 
#     Min. no. of genes annotated to a GO: 5 
#     Nontrivial nodes: 9106 
GSEA_BPres <- GenTable(GSEAGOdata_BP,
                       weight01KS = KSres,
                       orderBy = "weight01KS",
                       ranksOf = "weight01KS",
                       topNodes = 203)
save(GSEA_BPres, file = "topGO_gseaGO_BPresult.Rdata")
```

![](https://upload-images.jianshu.io/upload_images/14383117-c592b9801d4647d3.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

## ç»“æœå¯¹æ¯”

å…ˆæ‰¾å‡ºä¸‰ç§å·¥å…·ç»“æœä¸­å…±æœ‰çš„ GO term:

```R
nrow(GOrillares)  ## GOrilla
# [1] 100
nrow(gseaGO_BPresult2)  ## clusterProfiler
# [1] 457
nrow(GSEA_BPres)  ## topGO
# [1] 203
```

```R
GOrillavscp <- intersect(GOrillares$GO.term, gseaGO_BPresult2$ID)
GOrillavstop <- intersect(GOrillares$GO.term, GSEA_BPres$GO.ID)
triple <- intersect(GOrillavscp, GOrillavstop)
triple
# [1] "GO:0050867"
```

åªæœ‰ä¸€ä¸ªğŸ˜³

```R
row.names(GOrillares) <- GOrillares$GO.term
row.names(GSEA_BPres) <- GSEA_BPres$GO.ID
theonly <- cbind(GOrillares[triple,],
                 gseaGO_BPresult2[triple,],
                 GSEA_BPres[triple,])
finetable <- subset(theonly, select = c("Description",
                    "Enrichment..N..B..n..b.", "P.value","FDR.q.value",
                    "setSize","enrichmentScore","pvalue", "p.adjust","qvalues", 
                    "Annotated", "Significant","weight01KS"))
finetable
#                                       Description Enrichment..N..B..n..b.
# GO:0050867 positive regulation of cell activation  35.01 (17224,328,12,8)
#             P.value FDR.q.value setSize enrichmentScore     pvalue
# GO:0050867 7.67e-11    6.28e-08     312       0.2948657 0.00203252
#              p.adjust    qvalues Annotated Significant weight01KS
# GO:0050867 0.04044123 0.03411746       313         174    0.00383
```

![](https://upload-images.jianshu.io/upload_images/14383117-c70247bdb0872439.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

çœ‹èµ·æ¥ï¼Œ *GOrilla* çš„ç»“æœå’Œå¸¸ç”¨çš„ä¸¤ä¸ªRåŒ…çš„ç»“æœå·®çš„è¿˜æ˜¯æŒºå¤šçš„ğŸ˜²

### å‡ºäºå¥½å¥‡çœ‹ä¸€ä¸‹ topGO å’Œ clusterProfiler ç»“æœçš„å·®åˆ«

```R
topvscp <- intersect(GSEA_BPres$GO.ID, gseaGO_BPresult2$ID)
length(topvscp)
# [1] 19
```

![](https://upload-images.jianshu.io/upload_images/14383117-c8e82831b0d6fe69.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

å¤§æ¦‚å°±æ˜¯è¿™ä¸ªäºšå­ã€‚

---
# æœ€åï¼Œå‘å¤§å®¶éš†é‡æ¨èç”Ÿä¿¡æŠ€èƒ½æ ‘çš„ä¸€ç³»åˆ—å¹²è´§ï¼

1.  ç”Ÿä¿¡æŠ€èƒ½æ ‘å…¨çƒå…¬ç›Šå·¡è®²ï¼š[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  Bç«™å…¬ç›Š74å°æ—¶ç”Ÿä¿¡å·¥ç¨‹å¸ˆæ•™å­¦è§†é¢‘åˆè¾‘ï¼š[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  æ‹›å­¦å¾’ï¼š[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)
