> æœ¬æ–‡æ˜¯å¯¹å‰å‡ å¤©çš„[æ‰“è„¸](https://www.jianshu.com/p/66cb28d9d924)
æ‰“è„¸ä¹‹å¤„å°±åœ¨äºå‘ç°äº†muscleçš„RåŒ…ï¼Œä»æ­¤ä¸ç”¨å†ä¸ºåºåˆ—æ¯”å¯¹ä¸œå¥”è¥¿è·‘â€”â€”
![](https://upload-images.jianshu.io/upload_images/14383117-1aefb07b37b09e46.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)




## Step 1. ä»åŸºå› åæ‹¿åˆ°åºåˆ—

![](https://upload-images.jianshu.io/upload_images/14383117-08ef1c8adf83805c.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

```R
sub1 <- c("HDAC1","HDAC2","HDAC3","HDAC8")
sub2 <- c("HDAC4","HDAC5","HDAC6","HDAC7","HDAC9","HDAC10")
sub3 <- paste0("SIRT",1:7)
sub4 <- "HDAC11"
allsymb <- c(sub1,sub2,sub3,sub4)
```

è¿™ä¸€æ­¥ç”¨åˆ°çš„åŒ…ï¼š`biomaRt` ï¼Œé€‰æ‹©ç”¨symbolè·å–è›‹ç™½è´¨åºåˆ—ã€‚

```R
library(biomaRt)
mart <- useMart("ensembl","hsapiens_gene_ensembl")
all_ppseqs <- getSequence(id = allsymb,
                          type="hgnc_symbol",
                          seqType="peptide",
                          mart = mart)
```

æ£€æŸ¥ä¸€ä¸‹ï¼Œå‘ç°å°‘äº†SIRT6.

```R
table(all_ppseqs$hgnc_symbol)
# 
#  HDAC1 HDAC10 HDAC11  HDAC2  HDAC3  HDAC4  HDAC5  HDAC6  HDAC7  HDAC8  HDAC9  SIRT1  SIRT2 
#      3      6     16     12      4      8      8     18     23     45     18      4     14 
#  SIRT3  SIRT4  SIRT5  SIRT7 
#     12      3      7      4 
```

ä¼šå‡ºç°éƒ¨åˆ†åŸºå› ååŒ¹é…ä¸åˆ°åºåˆ—çš„æƒ…å†µï¼Œä¸€èˆ¬ç”¨ ENTREZ ID å¯ä»¥é¿å…è¿™ç§æƒ…å†µã€‚äºæ˜¯ç°åœ¨éœ€è¦è¡¥ä¸ŠSIRT6çš„åºåˆ—ã€‚

```R
library(org.Hs.eg.db)
sirt6entrez <- AnnotationDbi::select(org.Hs.eg.db,"SIRT6","ENTREZID","SYMBOL")
sirt6entrez
#   SYMBOL ENTREZID
# 1  SIRT6    51548
sirt6_ppseqs <- getSequence(id = "51548",
                            type="entrezgene_id",
                            seqType="peptide",
                            mart = mart)
```

å†å’Œå…¶ä»–åºåˆ—åˆå¹¶æˆä¸€ä¸ªæ•°æ®æ¡†ï¼Œå»é™¤unavailableåºåˆ—ã€‚

![](https://upload-images.jianshu.io/upload_images/14383117-c0fed46bd98ffaa6.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

```R
colnames(sirt6_ppseqs)[2] <- "hgnc_symbol" 
sirt6_ppseqs[,2] <- "SIRT6"
allppseqs <- rbind(all_ppseqs,sirt6_ppseqs)
allppseqs <- allppseqs[allppseqs$peptide!="Sequence unavailable",]
allppseqs <- allppseqs[order(allppseqs$hgnc_symbol),]
```

ç»™é‡å¤çš„åŸºå› åé‡å‘½åï¼š

```r
index <- duplicated(allppseqs$hgnc_symbol)
i <- 1
for(j in 1:nrow(allppseqs)){
if(index[j]==FALSE){
  i = 1
  allppseqs[j,2] <- paste0(allppseqs[j,2], ".",i)
  }
  else {
    i = i+1
    allppseqs[j,2] <- paste0(allppseqs[j,2], ".",i)
  } 
}
```

![](https://upload-images.jianshu.io/upload_images/14383117-d75cdc9f8e740d3b.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

è¾“å‡ºä¸ºFASTAæ–‡ä»¶ï¼š

```R
exportFASTA(allppseqs,"allppseqs.fasta")
```

## Step 2. å¤šåºåˆ—æ¯”å¯¹åŠç»˜åˆ¶è¿›åŒ–æ ‘

ç”¨åˆ°çš„åŒ…ï¼š`Biostrings` `muscle` `ape`

```R
library(Biostrings)
library(muscle)
library(ape)
```
MUSCLE ç®—æ³•çš„ç‰¹ç‚¹ä¹‹ä¸€å°±æ˜¯å¿«ï¼Œé‡‡ç”¨äº†k-merçš„å…¨å±€/æˆå¯¹æ¯”å¯¹æ–¹æ³•ï¼Œæœ€åå¾—å‡ºè®¡åˆ†çŸ©é˜µã€‚
MUSCLE åœ¨EBIçš„ç½‘é¡µç‰ˆå·¥å…·ä¼ é€é—¨ğŸ‘‰ https://www.ebi.ac.uk/Tools/msa/muscle/
MUSCLE è½¯ä»¶ç‰ˆä¼ é€é—¨ğŸ‘‰ http://www.drive5.com/muscle/downloads.htm

> The MUSCLE algorithm is a progressive alignment method that works with DNA, RNA, and amino acidsequences producing high-accuracy alignments with very fast computational times (Edgar, 2004,a). The algorithm is iterative, with later iterations refining the earlier alignments. In each iteration, pairwise alignment scores are computed for all sequence pairs (based on k-mer counting or global pairwise alignments) and the values are entered into a triangular distance matrix. 

è¯»å…¥FASTAæ–‡ä»¶ï¼š

```R
myseq <- readAAStringSet("allppseqs.fasta", format="fasta",
                          nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)
```

![](https://upload-images.jianshu.io/upload_images/14383117-6182cc3648f3aa4d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

muscleå¤šåºåˆ—æ¯”å¯¹ï¼š

```R
aln <- muscle::muscle(myseq)
```

è¿™ä¸€æ­¥ä¸€èˆ¬æ¯”è¾ƒè´¹æ—¶ã€‚

![](https://upload-images.jianshu.io/upload_images/14383117-1b37e160309d04c1.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

å½“ç„¶ `muscle()` æœ‰å¾ˆå¤šå‚æ•°å¯ä»¥è®¾ç½®ï¼Œæé€Ÿæˆ–è®¾å®šç©ºä½ç½šåˆ†ç­‰ç­‰ï¼š

- `diags = TRUE` é€Ÿåº¦æé«˜è‡ªç„¶æ„å‘³ç€å‡†ç¡®åº¦ä¸‹é™

  > **Enhanced speed**. To enhance the speed of the algorithm, the diags = TRUE flag will optimize the speed with a potential loss of accuracy.

- `gapopen = -30` ç©ºä½ç½šåˆ†

  > **Gap penalties**. Default gap penalties can be modified to produce altered alignments. The gap penalty must be negative, with larger negative values indicating more stringent penalties.

- ` maxhours = 24.0` æ¯”å¯¹è¿è¡Œçš„æœ€å¤§æ—¶é™

  > **Maximum number of hours**. If an alignment is expected to take a long time, a maximum total numberof hours can be specified, which, if reached, will lead to the algorithm stopping at this point and returningthe current alignment.

```R
aln
detail(aln) 
```

![](https://upload-images.jianshu.io/upload_images/14383117-28d09d66dcc18118.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![](https://upload-images.jianshu.io/upload_images/14383117-86196a800dd6a588.gif?imageMogr2/auto-orient/strip)

ä¿®å‰ªåºåˆ—ï¼š

```R
auto <- maskGaps(aln, min.fraction=0.5, min.block.width=4)
auto
```

é™¤æ­¤ä¹‹å¤–è¿˜å¯ä»¥ç”¨ `rowmask()` `colmask()` æ‰‹åŠ¨è®¾ç½®maskingåŒºåŸŸã€‚

![](https://upload-images.jianshu.io/upload_images/14383117-e6349258a5b76233.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![](https://upload-images.jianshu.io/upload_images/14383117-8076ec6168aad360.gif?imageMogr2/auto-orient/strip)

**ç”»æ ‘ï¼**

```R
sdist <- stringDist(as(aotu,"AAStringSet"), method="hamming") 
clust <- hclust(sdist, method = "single")
plot(as.phylo(clust),type="fan",cex = 0.8) 
```

![](https://upload-images.jianshu.io/upload_images/14383117-7f3381ce151e9c65.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

## References
- A guide to using muscle https://bioconductor.org/packages/release/bioc/vignettes/muscle/inst/doc/muscle-vignette.pdf
- plotå•ç‹¬ç”»å‡ºpheatmapè¿”å›çš„èšç±»ç»“æœï¼ˆèšç±»æ ‘ï¼‰[http://www.omicsclass.com/article/510](http://www.omicsclass.com/article/510)

---
# æœ€åï¼Œå‘å¤§å®¶éš†é‡æ¨èç”Ÿä¿¡æŠ€èƒ½æ ‘çš„ä¸€ç³»åˆ—å¹²è´§ï¼

1.  ç”Ÿä¿¡æŠ€èƒ½æ ‘å…¨çƒå…¬ç›Šå·¡è®²ï¼š[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  Bç«™å…¬ç›Š74å°æ—¶ç”Ÿä¿¡å·¥ç¨‹å¸ˆæ•™å­¦è§†é¢‘åˆè¾‘ï¼š[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  æ‹›å­¦å¾’ï¼š[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)


