> 只要你足够背，就没有烦恼可以追上你
毕竟，人生总是起起伏伏伏伏伏
After all, modern problem requires modern solution.
![不要再强求别人开心了](https://upload-images.jianshu.io/upload_images/14383117-8329974fbd0e5579.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

#### 初衷是想一切都在R里搞定来着
用到的包：`biomaRt` `Biostrings`, 之前的博客分别介绍过这两个包，参考价值嘛，一[般、]([https://www.jianshu.com/p/1e342a327914](https://www.jianshu.com/p/1e342a327914)
)[般]([https://www.jianshu.com/p/658c27e69466](https://www.jianshu.com/p/658c27e69466)
)

用到的软件：`ClustalW`/`ClustalX`, 当然可以用其他各种代替


## 先由基因名拿到序列

确定要用的数据框和数据集：

```R
library(biomaRt)
mart <- useMart("ensembl","hsapiens_gene_ensembl")
```

给自己的基因名构建一个向量：

```R
symbls <- c( "HHLA3","SCHLAP1","HHLA2","HLA-F","HLA-F-AS1","HLA-H",
            "HLA-T","HLA-K","HLA-U","HLA-A","HLA-E","HLA-C",
            "HLA-B","HLA-DRA","HLA-DRB5","HLA-DRB1","HLA-DQA1","HLA-DQB1",
            "HLA-DQB1-AS1","HLA-DMB","HLA-DMA","HHLA1")
length(symbls)
# [1] 22
```

前戏结束。

这时候需要确定想得到的序列类型：

> - *cdna*
> - *peptide* for protein sequences
> - *3utr* for 3’ UTR sequences
> - *5utr* for 5’ UTR sequences
> - *gene_exon* for exon sequences only
> - *transcript_exon* for transcript specific exonic sequences only
> - *transcript_exon_intron* gives the full unspliced transcript, that is exons + introns
> - *gene_exon_intron* gives the exons + introns of a gene
> - *coding* gives the coding sequence only
> - *coding_transcript_flank* gives the flanking region of the transcript including the UTRs, this must be accompanied with a given value for the upstream or downstream attribute
> - *coding_gene_flank* gives the flanking region of the gene including the UTRs, this must be accompanied with a given value for the upstream or downstream attribute
> - *transcript_flank* gives the flanking region of the transcript exculding the UTRs, this must be accompanied with a given value for the upstream or downstream attribute
> - *gene_flank* gives the flanking region of the gene excluding the UTRs, this must be accompanied with a given value for the upstream or downstream attribute

鉴于只有symbol，同一个symbol拿到的序列肯定很多，为了减少运算量选择结果最少的，经过试验 `gene_exon_intron` 是不多不少且全面~~显得比较负责~~的一个参数。实际情况还要看具体需求（主要是我么得需求😑

```R
gene_seqs <- getSequence(id = symbls,
                         type="hgnc_symbol",
                         seqType="gene_exon_intron",
                         mart = mart)
## 排一下序
gene_seqs <- gene_seqs[order(gene_seqs$hgnc_symbol),]
## 去重，只好残忍留下第一个，去掉其他所有
duplicated(gene_seqs$hgnc_symbol)
#   [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE
#  [13]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
#  [25] FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
#  [37]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE
#  [49]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
#  [61]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE  TRUE
#  [73]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
#  [85]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
#  [97]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
# [109]  TRUE  TRUE FALSE
gene_seqs <- gene_seqs[!index,]
## 完美解决
table(gene_seqs$hgnc_symbol)
# 
#        HHLA1        HHLA2        HHLA3        HLA-A        HLA-B        HLA-C 
#            1            1            1            1            1            1 
#      HLA-DMA      HLA-DMB     HLA-DQA1     HLA-DQB1 HLA-DQB1-AS1      HLA-DRA 
#            1            1            1            1            1            1 
#     HLA-DRB1     HLA-DRB5        HLA-E        HLA-F    HLA-F-AS1        HLA-H 
#            1            1            1            1            1            1 
#        HLA-K        HLA-T        HLA-U      SCHLAP1 
#            1            1            1            1 
```

然后就可以导出为fasta文件了：

```R
exportFASTA(gene_seqs,"symb2seq.fasta")
```

![](https://upload-images.jianshu.io/upload_images/14383117-5d8a67c98d41be43.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

天真地直接读，想直接比对：

```R
library(Biostrings)
myseq <- readDNAStringSet("symb2seq2.fasta", format="fasta",
                          nrec=-1L, skip=0L, 
                          seek.first.rec=FALSE, use.names=TRUE)
aligntest <- DNAMultipleAlignment(myseq)
# Error in validObject(.Object) : 
#   invalid class “DNAMultipleAlignment” object: alignments have an unequal number of characters
```

hmmmm...所以也并不能直接比对，比对还是要依赖其他软件，`Biostring` 只能对已经比对好的文件进行masking, 最后画树。

以下是各种无用尝试：

ClustalX比对即无响应

![](https://upload-images.jianshu.io/upload_images/14383117-6e20098590028178.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

muscle终于搭梯子下载好后点击即闪退

终于还是用了ClustalW

![](https://upload-images.jianshu.io/upload_images/14383117-ed6cf46266b1e44b.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![](https://upload-images.jianshu.io/upload_images/14383117-4722b7d5d4d2b2a8.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

hmmmm 一个多小时后


![](https://upload-images.jianshu.io/upload_images/14383117-5860bfb442550d5d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


下面是去睡觉的分割线

---
一觉醒来毫无结果文件输出🙃

![](https://upload-images.jianshu.io/upload_images/14383117-06a9f8c7ff952a05.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

于是又重、新、来、过 
几个小时之后，同样的情景再次出现，但神奇的是，昨天打不开的muscle网页版终于顺利打开了
![](https://upload-images.jianshu.io/upload_images/14383117-0b868be2c13b281f.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
又过了几个小时
![](https://upload-images.jianshu.io/upload_images/14383117-3e3ab4e0e364fc03.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
这就不得不怀疑序列本身是不是有问题，毕竟是随便丢了几个symbol得到的序列，额且：
```R
myseq <- readDNAStringSet("symb2seq.fasta", format="fasta",
                  nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)
myseq@ranges
# group  start    end  width        names
# 1      1      1  49674  49674        HHLA1
# 2      1  49675 131470  81796        HHLA2
# 3      1 131471 162005  30535        HHLA3
# 4      1 162006 282329 120324        HLA-A
# 5      1 282330 370307  87978        HLA-B
# 6      1 370308 373689   3382        HLA-C
# 7      1 373690 394171  20482      HLA-DMA
# 8      1 394172 412669  18498      HLA-DMB
# 9      1 412670 431553  18884     HLA-DQA1
# 10     1 431554 439165   7612     HLA-DQB1
# 11     1 439166 440015    850 HLA-DQB1-AS1
# 12     1 440016 445184   5169      HLA-DRA
# 13     1 445185 456264  11080     HLA-DRB1
# 14     1 456265 469199  12935     HLA-DRB5
# 15     1 469200 473938   4739        HLA-E
# 16     1 473939 489682  15744        HLA-F
# 17     1 489683 512121  22439    HLA-F-AS1
# 18     1 512122 514852   2731        HLA-H
# 19     1 514853 517627   2775        HLA-K
# 20     1 517628 518759   1132        HLA-T
# 21     1 518760 518945    186        HLA-U
# 22     1 518946 743781 224836      SCHLAP1
```
各个长度差了那么多，强行全局比对软件应该是想拒绝的。
那就用同一个symbol的不同序列做比对吧 (既然已经逐渐偏离初衷...
```R
HLAK_seqs <- gene_seqs2[gene_seqs2$hgnc_symbol=="HLA-K",]
for (i in 1:nrow(HLAK_seqs)) {
  HLAK_seqs[i,2] <- paste0(HLAK_seqs[i,2],".",i)
}
exportFASTA(HLAK_seqs,"HLAKseq.fasta")
```
ClustalX比对瞬间成功
```r
HLAKalign <- readDNAMultipleAlignment(filepath = "HLAKseq.aln",format = "clustal")
HLAKalign
# DNAMultipleAlignment with 6 rows and 2783 columns
#      aln                                                   names               
#  [1] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.1
#  [2] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.4
#  [3] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.5
#  [4] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCATGGCTTGTAAA HLA-K.2
#  [5] ATGGGGTCATGGCGTCCCAAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.3
#  [6] ATGGGGTTATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.6
```
```r
detail(HLAKalign) 
```
![](https://upload-images.jianshu.io/upload_images/14383117-e68e565f56afdbdf.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

"TATA" mask 和 auto-mask
```r
tataMasked <- maskMotif(HLAKalign, "TATA") 
autoMasked <- maskGaps(HLAKalign, min.fraction=0.5, min.block.width=4) 
full <- as.matrix(HLAKalign) 
dim(full)
# [1]    6 2783
partialta <- as.matrix(tataMasked)
dim(partialta)
# [1]    6 2779
partial <- as.matrix(autoMasked) 
dim(partial)
# [1]    6 2783
```
查看碱基频率和共有序列
```R
alphabetFrequency(tataMasked)
#        A   C   G   T M R W S Y K V H D B N - + .
# [1,] 555 783 853 580 0 0 0 0 0 0 0 0 0 0 0 8 0 0
# [2,] 562 774 851 587 0 0 0 0 0 0 0 0 0 0 0 5 0 0
# [3,] 551 787 853 579 0 0 0 0 0 0 0 0 0 0 0 9 0 0
# [4,] 563 773 850 588 0 0 0 0 0 0 0 0 0 0 0 5 0 0
# [5,] 558 782 848 582 0 0 0 0 0 0 0 0 0 0 0 9 0 0
# [6,] 556 783 852 581 0 0 0 0 0 0 0 0 0 0 0 7 0 0
consensusViews(tataMasked)
#   Views on a 2783-letter BString subject
# subject: ATGGGGTCATGGCGTCCCGAACCCTCCT...TCTGATGTGTCTCTCACGGCTTGTAAA
# views:
#     start  end width
# [1]     1 1638  1638 [ATGGGGTCATGGCGTCCCGAA...TCCACCATTCCRTCTCTGAC]
# [2]  1643 2783  1141 [AGGCCACCCTGAGGTGCTGGG...TGTCTCTCACGGCTTGTAAA]
```
画树：
```r
## 需要转换object类型
class(tataMasked)
# [1] "DNAMultipleAlignment"
# attr(,"package")
# [1] "Biostrings"
sdist <- stringDist(as(tataMasked,"DNAStringSet"), method="hamming") 
clust <- hclust(sdist, method = "single")
plot(clust) 
> 
```
![](https://upload-images.jianshu.io/upload_images/14383117-c162377eded38f33.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

所以目前看起来，依然是需要依靠软件进行比对o_o ....

---
# 最后，向大家隆重推荐生信技能树的一系列干货！

1.  生信技能树全球公益巡讲：[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  B站公益74小时生信工程师教学视频合辑：[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  招学徒：[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)
