> åªè¦ä½ è¶³å¤ŸèƒŒï¼Œå°±æ²¡æœ‰çƒ¦æ¼å¯ä»¥è¿½ä¸Šä½ 
æ¯•ç«Ÿï¼Œäººç”Ÿæ€»æ˜¯èµ·èµ·ä¼ä¼ä¼ä¼ä¼
After all, modern problem requires modern solution.
![ä¸è¦å†å¼ºæ±‚åˆ«äººå¼€å¿ƒäº†](https://upload-images.jianshu.io/upload_images/14383117-8329974fbd0e5579.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

#### åˆè¡·æ˜¯æƒ³ä¸€åˆ‡éƒ½åœ¨Ré‡Œæå®šæ¥ç€
ç”¨åˆ°çš„åŒ…ï¼š`biomaRt` `Biostrings`, ä¹‹å‰çš„åšå®¢åˆ†åˆ«ä»‹ç»è¿‡è¿™ä¸¤ä¸ªåŒ…ï¼Œå‚è€ƒä»·å€¼å˜›ï¼Œä¸€[èˆ¬ã€]([https://www.jianshu.com/p/1e342a327914](https://www.jianshu.com/p/1e342a327914)
)[èˆ¬]([https://www.jianshu.com/p/658c27e69466](https://www.jianshu.com/p/658c27e69466)
)

ç”¨åˆ°çš„è½¯ä»¶ï¼š`ClustalW`/`ClustalX`, å½“ç„¶å¯ä»¥ç”¨å…¶ä»–å„ç§ä»£æ›¿


## å…ˆç”±åŸºå› åæ‹¿åˆ°åºåˆ—

ç¡®å®šè¦ç”¨çš„æ•°æ®æ¡†å’Œæ•°æ®é›†ï¼š

```R
library(biomaRt)
mart <- useMart("ensembl","hsapiens_gene_ensembl")
```

ç»™è‡ªå·±çš„åŸºå› åæ„å»ºä¸€ä¸ªå‘é‡ï¼š

```R
symbls <- c( "HHLA3","SCHLAP1","HHLA2","HLA-F","HLA-F-AS1","HLA-H",
            "HLA-T","HLA-K","HLA-U","HLA-A","HLA-E","HLA-C",
            "HLA-B","HLA-DRA","HLA-DRB5","HLA-DRB1","HLA-DQA1","HLA-DQB1",
            "HLA-DQB1-AS1","HLA-DMB","HLA-DMA","HHLA1")
length(symbls)
# [1] 22
```

å‰æˆç»“æŸã€‚

è¿™æ—¶å€™éœ€è¦ç¡®å®šæƒ³å¾—åˆ°çš„åºåˆ—ç±»å‹ï¼š

> - *cdna*
> - *peptide* for protein sequences
> - *3utr* for 3â€™ UTR sequences
> - *5utr* for 5â€™ UTR sequences
> - *gene_exon* for exon sequences only
> - *transcript_exon* for transcript specific exonic sequences only
> - *transcript_exon_intron* gives the full unspliced transcript, that is exons + introns
> - *gene_exon_intron* gives the exons + introns of a gene
> - *coding* gives the coding sequence only
> - *coding_transcript_flank* gives the flanking region of the transcript including the UTRs, this must be accompanied with a given value for the upstream or downstream attribute
> - *coding_gene_flank* gives the flanking region of the gene including the UTRs, this must be accompanied with a given value for the upstream or downstream attribute
> - *transcript_flank* gives the flanking region of the transcript exculding the UTRs, this must be accompanied with a given value for the upstream or downstream attribute
> - *gene_flank* gives the flanking region of the gene excluding the UTRs, this must be accompanied with a given value for the upstream or downstream attribute

é‰´äºåªæœ‰symbolï¼ŒåŒä¸€ä¸ªsymbolæ‹¿åˆ°çš„åºåˆ—è‚¯å®šå¾ˆå¤šï¼Œä¸ºäº†å‡å°‘è¿ç®—é‡é€‰æ‹©ç»“æœæœ€å°‘çš„ï¼Œç»è¿‡è¯•éªŒ `gene_exon_intron` æ˜¯ä¸å¤šä¸å°‘ä¸”å…¨é¢~~æ˜¾å¾—æ¯”è¾ƒè´Ÿè´£~~çš„ä¸€ä¸ªå‚æ•°ã€‚å®é™…æƒ…å†µè¿˜è¦çœ‹å…·ä½“éœ€æ±‚ï¼ˆä¸»è¦æ˜¯æˆ‘ä¹ˆå¾—éœ€æ±‚ğŸ˜‘

```R
gene_seqs <- getSequence(id = symbls,
                         type="hgnc_symbol",
                         seqType="gene_exon_intron",
                         mart = mart)
## æ’ä¸€ä¸‹åº
gene_seqs <- gene_seqs[order(gene_seqs$hgnc_symbol),]
## å»é‡ï¼Œåªå¥½æ®‹å¿ç•™ä¸‹ç¬¬ä¸€ä¸ªï¼Œå»æ‰å…¶ä»–æ‰€æœ‰
duplicated(gene_seqs$hgnc_symbol)
#   [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE
#  [13]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
#  [25] FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
#  [37]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE
#  [49]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
#  [61]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE  TRUE
#  [73]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
#  [85]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
#  [97]  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
# [109]  TRUE  TRUE FALSE
gene_seqs <- gene_seqs[!index,]
## å®Œç¾è§£å†³
table(gene_seqs$hgnc_symbol)
# 
#        HHLA1        HHLA2        HHLA3        HLA-A        HLA-B        HLA-C 
#            1            1            1            1            1            1 
#      HLA-DMA      HLA-DMB     HLA-DQA1     HLA-DQB1 HLA-DQB1-AS1      HLA-DRA 
#            1            1            1            1            1            1 
#     HLA-DRB1     HLA-DRB5        HLA-E        HLA-F    HLA-F-AS1        HLA-H 
#            1            1            1            1            1            1 
#        HLA-K        HLA-T        HLA-U      SCHLAP1 
#            1            1            1            1 
```

ç„¶åå°±å¯ä»¥å¯¼å‡ºä¸ºfastaæ–‡ä»¶äº†ï¼š

```R
exportFASTA(gene_seqs,"symb2seq.fasta")
```

![](https://upload-images.jianshu.io/upload_images/14383117-5d8a67c98d41be43.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

å¤©çœŸåœ°ç›´æ¥è¯»ï¼Œæƒ³ç›´æ¥æ¯”å¯¹ï¼š

```R
library(Biostrings)
myseq <- readDNAStringSet("symb2seq2.fasta", format="fasta",
                          nrec=-1L, skip=0L, 
                          seek.first.rec=FALSE, use.names=TRUE)
aligntest <- DNAMultipleAlignment(myseq)
# Error in validObject(.Object) : 
#   invalid class â€œDNAMultipleAlignmentâ€ object: alignments have an unequal number of characters
```

hmmmm...æ‰€ä»¥ä¹Ÿå¹¶ä¸èƒ½ç›´æ¥æ¯”å¯¹ï¼Œæ¯”å¯¹è¿˜æ˜¯è¦ä¾èµ–å…¶ä»–è½¯ä»¶ï¼Œ`Biostring` åªèƒ½å¯¹å·²ç»æ¯”å¯¹å¥½çš„æ–‡ä»¶è¿›è¡Œmasking, æœ€åç”»æ ‘ã€‚

ä»¥ä¸‹æ˜¯å„ç§æ— ç”¨å°è¯•ï¼š

ClustalXæ¯”å¯¹å³æ— å“åº”

![](https://upload-images.jianshu.io/upload_images/14383117-6e20098590028178.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

muscleç»ˆäºæ­æ¢¯å­ä¸‹è½½å¥½åç‚¹å‡»å³é—ªé€€

ç»ˆäºè¿˜æ˜¯ç”¨äº†ClustalW

![](https://upload-images.jianshu.io/upload_images/14383117-ed6cf46266b1e44b.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![](https://upload-images.jianshu.io/upload_images/14383117-4722b7d5d4d2b2a8.jpg?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

hmmmm ä¸€ä¸ªå¤šå°æ—¶å


![](https://upload-images.jianshu.io/upload_images/14383117-5860bfb442550d5d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)


ä¸‹é¢æ˜¯å»ç¡è§‰çš„åˆ†å‰²çº¿

---
ä¸€è§‰é†’æ¥æ¯«æ— ç»“æœæ–‡ä»¶è¾“å‡ºğŸ™ƒ

![](https://upload-images.jianshu.io/upload_images/14383117-06a9f8c7ff952a05.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

äºæ˜¯åˆé‡ã€æ–°ã€æ¥ã€è¿‡ 
å‡ ä¸ªå°æ—¶ä¹‹åï¼ŒåŒæ ·çš„æƒ…æ™¯å†æ¬¡å‡ºç°ï¼Œä½†ç¥å¥‡çš„æ˜¯ï¼Œæ˜¨å¤©æ‰“ä¸å¼€çš„muscleç½‘é¡µç‰ˆç»ˆäºé¡ºåˆ©æ‰“å¼€äº†
![](https://upload-images.jianshu.io/upload_images/14383117-0b868be2c13b281f.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
åˆè¿‡äº†å‡ ä¸ªå°æ—¶
![](https://upload-images.jianshu.io/upload_images/14383117-3e3ab4e0e364fc03.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
è¿™å°±ä¸å¾—ä¸æ€€ç–‘åºåˆ—æœ¬èº«æ˜¯ä¸æ˜¯æœ‰é—®é¢˜ï¼Œæ¯•ç«Ÿæ˜¯éšä¾¿ä¸¢äº†å‡ ä¸ªsymbolå¾—åˆ°çš„åºåˆ—ï¼Œé¢ä¸”ï¼š
```R
myseq <- readDNAStringSet("symb2seq.fasta", format="fasta",
                  nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)
myseq@ranges
# group  start    end  width        names
# 1      1      1  49674  49674        HHLA1
# 2      1  49675 131470  81796        HHLA2
# 3      1 131471 162005  30535        HHLA3
# 4      1 162006 282329 120324        HLA-A
# 5      1 282330 370307  87978        HLA-B
# 6      1 370308 373689   3382        HLA-C
# 7      1 373690 394171  20482      HLA-DMA
# 8      1 394172 412669  18498      HLA-DMB
# 9      1 412670 431553  18884     HLA-DQA1
# 10     1 431554 439165   7612     HLA-DQB1
# 11     1 439166 440015    850 HLA-DQB1-AS1
# 12     1 440016 445184   5169      HLA-DRA
# 13     1 445185 456264  11080     HLA-DRB1
# 14     1 456265 469199  12935     HLA-DRB5
# 15     1 469200 473938   4739        HLA-E
# 16     1 473939 489682  15744        HLA-F
# 17     1 489683 512121  22439    HLA-F-AS1
# 18     1 512122 514852   2731        HLA-H
# 19     1 514853 517627   2775        HLA-K
# 20     1 517628 518759   1132        HLA-T
# 21     1 518760 518945    186        HLA-U
# 22     1 518946 743781 224836      SCHLAP1
```
å„ä¸ªé•¿åº¦å·®äº†é‚£ä¹ˆå¤šï¼Œå¼ºè¡Œå…¨å±€æ¯”å¯¹è½¯ä»¶åº”è¯¥æ˜¯æƒ³æ‹’ç»çš„ã€‚
é‚£å°±ç”¨åŒä¸€ä¸ªsymbolçš„ä¸åŒåºåˆ—åšæ¯”å¯¹å§ (æ—¢ç„¶å·²ç»é€æ¸åç¦»åˆè¡·...
```R
HLAK_seqs <- gene_seqs2[gene_seqs2$hgnc_symbol=="HLA-K",]
for (i in 1:nrow(HLAK_seqs)) {
  HLAK_seqs[i,2] <- paste0(HLAK_seqs[i,2],".",i)
}
exportFASTA(HLAK_seqs,"HLAKseq.fasta")
```
ClustalXæ¯”å¯¹ç¬é—´æˆåŠŸ
```r
HLAKalign <- readDNAMultipleAlignment(filepath = "HLAKseq.aln",format = "clustal")
HLAKalign
# DNAMultipleAlignment with 6 rows and 2783 columns
#      aln                                                   names               
#  [1] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.1
#  [2] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.4
#  [3] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.5
#  [4] ATGGGGTCATGGCGTCCCGAACCCT...TGATGTGTCTCTCATGGCTTGTAAA HLA-K.2
#  [5] ATGGGGTCATGGCGTCCCAAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.3
#  [6] ATGGGGTTATGGCGTCCCGAACCCT...TGATGTGTCTCTCACGGCTTGTAAA HLA-K.6
```
```r
detail(HLAKalign) 
```
![](https://upload-images.jianshu.io/upload_images/14383117-e68e565f56afdbdf.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

"TATA" mask å’Œ auto-mask
```r
tataMasked <- maskMotif(HLAKalign, "TATA") 
autoMasked <- maskGaps(HLAKalign, min.fraction=0.5, min.block.width=4) 
full <- as.matrix(HLAKalign) 
dim(full)
# [1]    6 2783
partialta <- as.matrix(tataMasked)
dim(partialta)
# [1]    6 2779
partial <- as.matrix(autoMasked) 
dim(partial)
# [1]    6 2783
```
æŸ¥çœ‹ç¢±åŸºé¢‘ç‡å’Œå…±æœ‰åºåˆ—
```R
alphabetFrequency(tataMasked)
#        A   C   G   T M R W S Y K V H D B N - + .
# [1,] 555 783 853 580 0 0 0 0 0 0 0 0 0 0 0 8 0 0
# [2,] 562 774 851 587 0 0 0 0 0 0 0 0 0 0 0 5 0 0
# [3,] 551 787 853 579 0 0 0 0 0 0 0 0 0 0 0 9 0 0
# [4,] 563 773 850 588 0 0 0 0 0 0 0 0 0 0 0 5 0 0
# [5,] 558 782 848 582 0 0 0 0 0 0 0 0 0 0 0 9 0 0
# [6,] 556 783 852 581 0 0 0 0 0 0 0 0 0 0 0 7 0 0
consensusViews(tataMasked)
#   Views on a 2783-letter BString subject
# subject: ATGGGGTCATGGCGTCCCGAACCCTCCT...TCTGATGTGTCTCTCACGGCTTGTAAA
# views:
#     start  end width
# [1]     1 1638  1638 [ATGGGGTCATGGCGTCCCGAA...TCCACCATTCCRTCTCTGAC]
# [2]  1643 2783  1141 [AGGCCACCCTGAGGTGCTGGG...TGTCTCTCACGGCTTGTAAA]
```
ç”»æ ‘ï¼š
```r
## éœ€è¦è½¬æ¢objectç±»å‹
class(tataMasked)
# [1] "DNAMultipleAlignment"
# attr(,"package")
# [1] "Biostrings"
sdist <- stringDist(as(tataMasked,"DNAStringSet"), method="hamming") 
clust <- hclust(sdist, method = "single")
plot(clust) 
> 
```
![](https://upload-images.jianshu.io/upload_images/14383117-c162377eded38f33.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

æ‰€ä»¥ç›®å‰çœ‹èµ·æ¥ï¼Œä¾ç„¶æ˜¯éœ€è¦ä¾é è½¯ä»¶è¿›è¡Œæ¯”å¯¹o_o ....

---
# æœ€åï¼Œå‘å¤§å®¶éš†é‡æ¨èç”Ÿä¿¡æŠ€èƒ½æ ‘çš„ä¸€ç³»åˆ—å¹²è´§ï¼

1.  ç”Ÿä¿¡æŠ€èƒ½æ ‘å…¨çƒå…¬ç›Šå·¡è®²ï¼š[https://mp.weixin.qq.com/s/E9ykuIbc-2Ja9HOY0bn_6g](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FE9ykuIbc-2Ja9HOY0bn_6g)
2.  Bç«™å…¬ç›Š74å°æ—¶ç”Ÿä¿¡å·¥ç¨‹å¸ˆæ•™å­¦è§†é¢‘åˆè¾‘ï¼š[https://mp.weixin.qq.com/s/IyFK7l_WBAiUgqQi8O7Hxw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FIyFK7l_WBAiUgqQi8O7Hxw)
3.  æ‹›å­¦å¾’ï¼š[https://mp.weixin.qq.com/s/KgbilzXnFjbKKunuw7NVfw](https://links.jianshu.com/go?to=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FKgbilzXnFjbKKunuw7NVfw)
