# Dragon star Day 1

> [Dragonstar2019 by Kai Wang](https://links.jianshu.com/go?to=https%3A%2F%2Fgithub.com%2FWGLab%2Fdragonstar2019)
>
> 1. Genomic technologies in disease studies
> 2. NGS data formats and variant calling

[TOC]

## Part â…  Genomic technologies in disease studies

### 1 Types of genetic variation

- **SNVs** (Single Nucleotide Variants) - å•æ ¸è‹·é…¸å˜å¼‚

- **Indel** (Insertion or deletion < 50 bp)

- **SV** (Structural Variants)  - ç»“æ„å˜å¼‚

  can be *balanced* or *unbalanced*
  
  - **Balanced** events do not involve gain or loss of genetic materials
  - Inversions and translocations
    - Complex SVs (several types together)
  
  - **Unbalanced** events:
  - deletions/insertions/duplications (Deletions and duplications are two subtypes of **CNVs** (Copy Number Variants)
    - Chromosomal aneuploidies (such as *trisomy 21* 21ä¸‰ä½“) are *extreme* cases of unbalanced SV.
  
- Allele frequency and effect size - ç­‰ä½åŸºå› é¢‘ç‡ä¸æ•ˆåº”é‡/æ•ˆåº”å€¼

    ![Manolio et al, Nature, 2009](https://s2.ax1x.com/2019/07/29/e8gcXq.png)
  
    > Genome-wide association studies (GWAS) are effective in detecting **common alleles** that contribute to the inherited component of **common multifactorial diseases**. Typically, the alleles identified by this approach have **modest effect sizes** that cannot fully account for disease susceptibility. This discrepancy may exist because it is hard to identify rare alleles with a low to modest penetrance using GWAS. Penetrance is a measure of the proportion of individuals in a population carrying a particular allele that expresses the related phenotype. In contrast to multifactorial diseases, Mendelian diseases have a high penetrance and very rare allele frequency.
    >
    > McCarthy, Mark I., et al. "Genome-wide association studies for complex traits: consensus, uncertainty and challenges." Nature reviews genetics 9.5 (2008): 356.
    
    > åœ¨[ç»Ÿè®¡å­¦](https://zh.wikipedia.org/wiki/çµ±è¨ˆå­¸)ä¸­ï¼Œ**æ•ˆåº”å€¼**ï¼ˆEffect sizeï¼‰æ˜¯é‡åŒ–ç°è±¡å¼ºåº¦çš„æ•°å€¼ã€‚[[1\]](https://zh.wikipedia.org/wiki/æ•ˆåº”å€¼#cite_note-Kelley2012-1)æ•ˆåº”å€¼å®é™…çš„ç»Ÿè®¡é‡åŒ…æ‹¬äº†äºŒä¸ªå˜æ•°é—´çš„[ç›¸å…³ç¨‹åº¦](https://zh.wikipedia.org/wiki/ç›¸é—œ)ã€å›å½’æ¨¡å‹ä¸­çš„[å›å½’ç³»æ•°](https://zh.wikipedia.org/wiki/ç·šæ€§å›æ­¸)ã€ä¸åŒå¤„ç†é—´[å¹³å‡å€¼](https://zh.wikipedia.org/wiki/å¹³å‡å€¼)çš„å·®å¼‚â€¦â€¦ç­‰ç­‰ã€‚æ— è®ºå“ªç§æ•ˆåº”å€¼ï¼Œå…¶[ç»å¯¹å€¼](https://zh.wikipedia.org/wiki/çµ•å°å€¼)è¶Šå¤§è¡¨ç¤ºæ•ˆåº”è¶Šå¼ºï¼Œä¹Ÿå°±æ˜¯ç°è±¡è¶Šæ˜æ˜¾ã€‚æ•ˆåº”å€¼ä¸[ç‰¹æ•ˆæ£€éªŒ](https://zh.wikipedia.org/wiki/å‡è¨­æª¢å®š)çš„æ¦‚å¿µæ˜¯äº’è¡¥çš„ã€‚åœ¨ä¼°ç®—[ç»Ÿè®¡æ£€å®šåŠ›](https://zh.wikipedia.org/wiki/çµ±è¨ˆæª¢å®šåŠ›)ã€éœ€è¦çš„[æ ·æœ¬æ•°](https://zh.wikipedia.org/w/index.php?title=æ¨£æœ¬æ•¸&action=edit&redlink=1)ä¸è¿›è¡Œ[å…ƒåˆ†æ](https://zh.wikipedia.org/wiki/å…ƒåˆ†æ)æ—¶ï¼Œæ•ˆåº”å€¼ç»å¸¸æ‰®æ¼”é‡è¦è§’è‰²ã€‚
    >
    > https://zh.wikipedia.org/wiki/æ•ˆåº”å€¼
    
    > Effect size is a simple way of quantifying the difference between two groups that has many advantages over the use of tests of statistical significance alone. Effect size emphasises the size of the difference rather than confounding this with sample size.* 
    >
    > Coe R. It's the effect size, stupid: What effect size is and why it is important[J]. 2002.

### 2 Revolution: single-molecule long-read sequencing 

å•åˆ†å­é•¿è¯»é•¿æµ‹åº

#### 2.1 Oxford Nanopore Sequencing 

æµ‹é‡ç”µæµ

The DNA/RNA is sequenced when it is going though a protein pore. 

- DNA (or RNA)
- Motor protein
- Adapter sequence

The nucleotides in the DNA/RNA block the **ionic current** and induce changes of current, which can be measured. 

#### 2.2 PacBio Single-molecule real-time (SMRT) sequencing

æµ‹é‡è§å…‰

- SMRTbell library: ç¯çŠ¶

- Adapter: ç¯çŠ¶

- Types of SMRT sequencing reads

  - Continuous Long Reads (CLR)

    > Long inserts so that the polymerase can synthesize along a single strand

  - Circular Consensus Sequencing (CCSï¼Œ HiFiï¼‰

    > Short inserts, so polymerase can continue around the entire SMRTbell multiple times and generate multiple sub-reads from the same single molecule

  - å­˜åœ¨ç”±äºæŠ€æœ¯é€ æˆçš„ error insertion
  
    Insertions tend to be more than deletions and substitutions.

#### 2.3 Linked-read Sequencing

ä¾é barcode

By adding a unique *barcode* to every short read generated from an individual molecule, the **short reads are linked together.**

> 10X Genomics å…¬å¸çš„ linked reads æŠ€è¡“æœ¬è³ªä¸Šæ˜¯å°‡ barcode åºåˆ—å¼•å…¥é•·åºåˆ—ç‰‡æ®µï¼Œé€éå°‡é•·ç‰‡æ®µåˆ†é…åˆ°ä¸åŒçš„æ²¹æ»´å¾®ç²’ä¸­ï¼Œåˆ©ç”¨ GemCode å¹³å°å°é•·ç‰‡æ®µåºåˆ—é€²è¡Œæ“´å¢å¼•å…¥ barcode åºåˆ—ä»¥åŠå®šåºæ¥é ­å¼•ç‰©ï¼Œç„¶å¾Œå°‡åºåˆ—æ‰“æ–·æˆé©åˆå®šåºå¤§å°çš„ç‰‡æ®µé€²è¡Œå®šåºï¼Œé€šé barcode åºåˆ—è³‡è¨Šè¿½è¸ªä¾†è‡ªæ¯å€‹å¤§ç‰‡æ®µDNAæ¨¡æ¿çš„å¤šå€‹ Readsï¼Œå¾è€Œç²å¾—å¤§ç‰‡æ®µçš„éºå‚³è³‡è¨Šã€‚é€šé linked reads çµåˆå¸¸è¦äºŒä»£å®šåºçµ„è£å¾—åˆ°çš„Scaffoldï¼Œå¯å»ºæ§‹æº–ç¢ºåº¦æ›´é•·çš„Scaffoldã€‚
>
> http://toolsbiotech.blog.fc2.com/blog-entry-37.html

> **What are molecular barcodes?**
> The concept of molecular barcodes is that each original DNA fragment, within the same sample, is attached to a unique sequence barcode. 
>
> <https://pdfs.semanticscholar.org/310b/3bac42989485c98406848217418ff22c22e7.pdf>

- Linked read use molecular barcoding to preserve long-range information

- Short read pairs (2 x150 bp) are generated using barcode-containing primers.
- Short reads that contains the **same barcode** and within a certain distance can be **linked together** to â€œreconstructâ€ the original long DNA fragment.

#### 2.4 Single-molecule optical mapping (Bionano Genomics) 

ä¾é é™åˆ¶æ€§å†…åˆ‡é…¶ä½ç‚¹ï¼Œæµ‹é‡å…‰è°±

> **Optical mapping** 
>
> By mapping the location of **restriction enzyme sites** along the unknown DNA of an organism, the spectrum of resulting DNA fragments collectively serves as a unique "fingerprint" or "barcode" for that sequence. 
>
> <https://en.wikipedia.org/wiki/Optical_mapping>

## Part â…¡ NGS data formats and variant calling

### 1 Basic Concepts in NGS

- Insert â€“ the DNA portion that is used for sequencing *note: å’Œçªå˜ä¸­çš„insertionä¸åŒ

  > Insert size is the length of the DNA (or RNA) that you want to sequence and that is "inserted" between the adapters (so adapters excluded).
  >
  > <https://www.biostars.org/p/95803/>

- Read â€“ the part of the insert that is sequenced è¯»é•¿

- Single End â€“ a sequencing procedure by which the insert is sequenced from one end only å•ç«¯

- Paired End â€“ a sequencing procedure by which the insert is sequenced from both ends åŒç«¯

### 2 Data formats

#### 2.1 The Rawset of Raw Data

**Typically: images** æ ¹æ®ä¸åŒç¢±åŸºå‘å°„å‡ºä¸åŒé¢œè‰²çš„è§å…‰

![](https://s2.ax1x.com/2019/08/06/e4iBMn.png)

**base calling**:  call nucleotides at each base of each read.

#### 2.2 NGS data formats

**FASTQ**: The raw sequence data format

Millions of short reads from unknown genetic locations.

![](https://s2.ax1x.com/2019/08/06/e4FpeP.png)

- Basic Qualities: a "Phred" scale

  *BQ* = -10log~10~(Îµ) where Îµ is the probability of an error.
  
  > Phred: *Phil's revised editor* by Phil Green
  >
  > https://www.pnas.org/content/101/39/13991>
  >
  
  > **Base quality conversion**
  >
  > Nowadays, we settled on using quality scores on the original Sanger format,(Phred+33).
  >
  > éœ€è¦çŸ¥é“æµ‹åºå¹´ä»½/å…¬å¸ï¼Œæ¥ç¡®å®šPhred
  >
  > ![](https://s2.ax1x.com/2019/07/31/eYAiUs.png)<https://en.wikipedia.org/wiki/FASTQ_format>
  
- **BED**: Genomic region format

  - The first 3 required BED fields are:
  
    1. **chrom**: "chr1"
  
    2. **chromStart**: "0"
  
    3. **chromEnd**: "3"
  
    åªåŒ…å«è¿™ä¸‰ç§ä¿¡æ¯çš„BED: Minimal BED format. So-called **BED3** format.
  
  - å†åŠ ä¸Šï¼š
  
    4. **labels**: "foo", "bar", "biz"...
    5. **scores**: "3.1"
    6. **strands**: "+", "-"
  
    This is so-called **BED6** format.
  
- **GFF**: annotates one line per feature

- **BAM/SAM**: Genome alignment format

  - SAM: Sequence Alignment/Map format (tab-delimited text file). 
  - BAM: The **binary** equivalent of a SAM file, which stores the same data in a compressed binary representation

  ![https://www.samformat.info/sam-format-flag](https://s2.ax1x.com/2019/08/07/e5ojrd.png)

- **CRAM**

  å’Œ BAM/SAM ç±»ä¼¼çš„å¦ä¸€ç§æ ¼å¼ï¼Œæ— æŸå‹ç¼©ï¼Œé€‚ç”¨äºå¤§å‹äººç¾¤åŸºå› ç»„/å¤–æ˜¾å­ç»„æµ‹åºé¡¹ç›®ï¼Œå¦‚ UK Biobank.

  - CRAM was designed to be an efficient reference-based **alternative** to SAM/BAM file formats
  - Better **lossless compression** than BAM, but also allow for controlled loss of BAM data

- **VCF**: Variant Call Format

  - A **gold** standard for describing variants.

  - One **locus** per line, and it may contain more than one mutations, but most lines contain **one variant only**.

  - ä¸»è¦åˆ†ä¸ºï¼š

    - the header line

      åŒ…æ‹¬ï¼šCHROM POS ID REF ALT QUAL FILTER INFO 

      å­˜åœ¨genotypeæ—¶ï¼Œè¿˜æœ‰ï¼šFORMAT

    - the INFO line

      ç”¨åˆ†å·åˆ†éš”çš„keys, ç”¨ç­‰å·å®šä¹‰å¯é€‰çš„å€¼

    - the genotypes

      å®šä¹‰æ•°æ®ç±»å‹åŠé¡ºåº

- **gVCF** (Genomic VCF): 

  - the basic format specification is **the same as for a regular VCF**, but gVCF contains **extra information**.
  - gVCFå‚¨å­˜äº†å˜å¼‚åŠéå˜å¼‚ä½ç‚¹çš„æµ‹åºä¿¡æ¯ï¼Œé€‚ç”¨äºä¸´åºŠåˆ†æ

#### 2.3 Formats use different coordinate systems, which adds confusion

ä¸åŒæ–‡ä»¶æ ¼å¼å¯¹æŸ“è‰²ä½“åæ ‡çš„èµ·å§‹ä½ç½®å®šä¹‰ä¸åŒï¼Œé€ æˆäº†ä¸€å®šçš„å›°æ‰°

> BED: 0-based, half-open
> GFF: 1-based, fully closed
> SAM: 1-based, fully closed
> BAM: 0-based, half-open
> VCF: 1-based, fully closed
> â€¦

### 3 Visualization of genomic data: IGV

å¼ºå¤§çš„IGV

**IGV**: a high-performance visualization tool for interactive exploration of large, integrated genomic datasets.

### 4 Coverage

#### 4.1 What is coverage?

- The **depth** of sequencing coverage can be defined theoretically as **LN/G**, where L is the **read length**, N is the **number of reads** and G is the **haploid genome length**.

  æµ‹åºæ·±åº¦å¯ç†è§£ä¸ºï¼š(readé•¿åº¦/readæ•°é‡)/å•å€ä½“åŸºå› ç»„é•¿åº¦

- The **breadth** of coverage is the **percentage of target bases** that have been sequenced for a given number of times.

- The **accuracy** of variant calling is affected by sequence quality, uniformity of coverage and the threshold of false-discovery rate that is used.

  > ## What is variant calling?
  >
  > Variant calling is the process by which we identify variants from sequence data (Figure 11).
  >
  > 1. Carry out whole genome or whole exome sequencing to create **FASTQ** files.
  > 2. Align the sequences to a reference genome, creating **BAM** or **CRAM** files.
  > 3. Identify where the aligned reads differ from the reference genome and write to a **VCF** file.
  >
  > Identify where the aligned reads differ from the reference genome and write to a VCF file.
  >
  > ![](https://www.ebi.ac.uk/training/online/sites/ebi.ac.uk.training.online/files/GenVar_Fig_CRAM_file.png)
  >
  > Figure 11 A CRAM file aligned to a reference genomic region as visualised in Ensembl. Differences are highlighted in red in the reads, and will be called as variants.
  >
  > readsä¸­æ ‡çº¢çš„ç¢±åŸºä¸ºvariants.
  >
  > <https://www.ebi.ac.uk/training/online/course/human-genetic-variation-i-introduction-2019/variant-identification-and-analysis>

#### 4.2 Coverage: how many reads we need to cover the genome?

- Depth of coverage model

  é¸Ÿæªæ³•æµ‹åºä¸­ï¼Œgenome size ***G***, read size ***S***, ***N*** reads

  æŸä¸ªreadå‡ºç°åœ¨æŸä¸ªé—´éš™(interval)çš„æ¦‚ç‡ä¸ºï¼š*p=L/G*

  ***D***: å‡ºç°åœ¨ä¸readåŒç­‰é•¿åº¦çš„é—´éš™çš„readçš„æ•°é‡

  *D* ~ Binomial(*N, L/G*)ï¼ŒDä¸N, p ç¬¦åˆäºŒé¡¹åˆ†å¸ƒ

  å³ï¼š**b(*x*; *n, P*) = ~n~C~x~* P^x^ * (1 - P) ^n-x^** 

  â€‹	   **b(*D*; *N,  L/G*) = ~N~C~D~* (L/G)^D^ * (1 - L/G) ^N-D^** 

  Let *L* = *S*, *D* is also the number of reads that cover the last position of the interval â†’ ***D* is depth of coverage**.

  å› ä¸º *S << G*, *N*çš„å€¼ç‰¹åˆ«å¤§ï¼Œdepthå¯è¿‘ä¼¼ç†è§£ä¸ºæ³Šæ¾åˆ†å¸ƒï¼š

  ***D ~ Poisson(Î»)***
  *Î»= SN/G* (average depth of coverage)

  $P(D=k)=\frac{e^{-\lambda}\lambda^k}{k!}$

- Fraction of genome that are covered

  æ ¹æ®ä¸Šè¿°å…¬å¼å¯ä»¥è®¡ç®—ï¼š

  Given Î»=40, the fraction of genome that are covered more than 30x (D>30) is: 0.938.

#### 4.3 Overdispersion

- Main cause of overdispersion

  **GC bias** and other technical factors lead to systematic bias in coverage, resulting in overdispersion.

- How to model overdispersion

  - Ideal situation (**Poisson distrition**): 
    $Var(D) = Î¼$

  - **Gamma-Poisson** is equivalent to Negative Binomial, which is a  commonly used model for dealing with overdispersion in count data: 
    $Var(D) = Î¼+ Î¼^2/k$

  - è¿‡ç¦»æ•£æ¨¡å‹ä¹Ÿé€‚ç”¨äºå…¶ä»–å› ç´ ï¼Œå¦‚ biological noise

#### 4.4 Question on coverage

Why do we need average 30-50x in a typical WGS experiment, and 100x in WES?

> WGS is less biased compared to WES. We do not need as much depth to call a variant confidently. Check the following publications for more detailed information.
>
> " Exome-seq achieves 95% SNP detection sensitivity at a mean on-target depth of 40 reads, whereas WGS only requires a mean of 14 reads. "
>
> Article [Variant detection sensitivity and biases in whole genome and...](https://www.researchgate.net/publication/264053300_Variant_detection_sensitivity_and_biases_in_whole_genome_and_exome_sequencing)
>
> [Shu-Hong Lin](https://www.researchgate.net/profile/Shu_Hong_Lin)
>
> <https://www.researchgate.net/post/Why_is_NGS_coverage_of_seemingly_less_complex_exome_sequences_higher_than_that_of_whole_genome_sequencing>

"WGS is less biased compared to WES."çš„åŸå› ï¼šWESçš„æ ·å“é¦–å…ˆéœ€è¦ç»è¿‡PCR

> ![](https://s2.ax1x.com/2019/08/07/eIJQhD.png)
>
> Why 30X WGS beats 100X WES for variant coverage
>
> <https://www.variantyx.com/variantyx-posts/why-30x-wgs-beats-100x-wes-for-variant-coverage/>

### 5 General strategy for variant calling

#### 5.1 Possible Genotypes

1. å½“åªæœ‰å‚è€ƒåŸºå› ç»„æ—¶ï¼Œå„ç§æƒ…å½¢çš„æ¦‚ç‡å‡ä¸º1

   ![](https://s2.ax1x.com/2019/08/07/eIodOJ.png)

   P(reads|A/A, read mapped) = P(C observed|A/A, read mapped) = 1.0

   P(reads|A/C, read mapped) = P(C observed|A/C, read mapped) = 1.0

   P(reads|C/C, read mapped) = P(C observed|C/C, read mapped) = 1.0

2. å‡å®šerror rateä¸º0.01

   ![](https://s2.ax1x.com/2019/08/07/eI7mrQ.png)

3. å½“ç¬¬1æ¡readçš„è¯¥ä½ç‚¹ä¸º**C**æ—¶

   P(reads|A/A, read mapped) = 0.01

   P(reads|A/C, read mapped) = 0.5

   P(reads|C/C, read mapped) = 1 - 0.01 = 0.99

3. å½“ç¬¬2æ¡readçš„è¯¥ä½ç‚¹ä¸º**C**æ—¶

   P(reads|A/A, read mapped) = 0.01^2^ = 0.0001

   P(reads|A/C, read mapped) = 0.5^2^ = 0.25

   P(reads|C/C, read mapped) = 0.99^2^ = 0.9801

4. å½“ç¬¬3æ¡readçš„è¯¥ä½ç‚¹ä¸º**C**æ—¶

   P(reads|A/A, read mapped) = 0.01^3^ = 0.000001

   P(reads|A/C, read mapped) = 0.5^3^ = 0.125

   P(reads|C/C, read mapped) = 0.99^3^ = 0.970299

5. å½“ç¬¬4æ¡readçš„è¯¥ä½ç‚¹ä¸º**A**æ—¶

   P(reads|A/A, read mapped) = 0.01^3^ * 0.99 = 0.00000099

   P(reads|A/C, read mapped) = 0.5^4^ = 0.0625

   P(reads|C/C, read mapped) = 0.99^3^ * 0.01 = 0.00970299

6. å½“ç¬¬5æ¡readçš„è¯¥ä½ç‚¹ä¸º**A**

   P(reads|A/A, read mapped) = 0.01^3^ * 0.99^2^ = **0.00000098**

   P(reads|A/C, read mapped) = 0.5^5^ = **0.03125**

   P(reads|C/C, read mapped) = 0.99^3^ * 0.01^2^ = 0.0000970299 â‰ˆ **0.000097**

æ€»ç»“å‡ºè´å¶æ–¯å…¬å¼ï¼š

![](https://s2.ax1x.com/2019/08/07/eI782T.png)

Combine these **likelihoods** with a prior incorporating information from other individuals and flanking sites to **assign a genotype**. 

#### 5.2 From Sequence to Genotype: Individual Based Prior

**Individual Based Prior**: Evry site has 1/1000 probability of varying.

> Ingredients That Go Into Prior
>
> - Most sites donâ€™t vary
>
>      P(non-reference base) ~ **0.001**
>
> - When a site does vary, it is usually heterozygous
>   P(non-reference heterozygote) ~ **0.001 * 2/3**
>   P(non-reference homozygote) ~ **0.001 * 1/3**
> - Mutation model
>   Transitions account for most variants (Câ†”T or Aâ†”G)
>   Transversions account for minority of variants
>
> <https://pdfs.semanticscholar.org/0156/04c3ba76cf247a8010f74ec1386e58ceb530.pdf>

å› æ­¤ï¼Œå„ä½ç‚¹çš„å…ˆéªŒæ¦‚ç‡ä¸ºï¼š

Prior(A/A) = 0.001 * 1/3 â‰ˆ  **0.00033**  0.00034?

Prior(A/C) = 0.001 * 2/3 â‰ˆ  **0.00067**  0.00066?

Prior(C/C) = 1 - 0.001 = **0.999**

ç¢ç£¨äº†ä¸¤ä¸ªå°æ—¶å¹¶æŠŠgoogleæœç´¢ç¿»åˆ°äº†ç¬¬åé¡µï¼Œä¾ç„¶æ²¡çœ‹æ‡‚è¿™é‡Œçš„åéªŒæ¦‚ç‡æ˜¯æ€ä¹ˆç®—çš„ï¼Œå¦‚æœæœ‰ç»Ÿè®¡å¤§ç¥çœ‹åˆ°è¿™é‡Œï¼Œæ±‚èµæ•™ğŸ™

![](https://s2.ax1x.com/2019/08/07/eokvkt.png)

#### 5.3 From Sequence to Genotype: Population Based Prior

**Population Based Prior**: Use frequency information from examining others at the same site. *In the example above, we estimated P(A) = 0.20*

å› æ­¤ï¼Œå„ä½ç‚¹çš„å…ˆéªŒæ¦‚ç‡ä¸ºï¼š

Prior(A/A) = 0.2 * 0.2 = **0.04**

Prior(A/C) = 1 - 0.04 - 0.64 = **0.32**

Prior(C/C) = 0.8 * 0.8 = **0.64**

åŒæ ·çš„è¿·æƒ‘ï¼š

![](https://s2.ax1x.com/2019/08/07/eoAKcF.png)

#### 5.4 Sequence Based Genotype Calls

> - **Individual Based Prior**
>   - Assumes all sites have an equal probability of showing polymorphism
>   - Specifically, assumption is that about 1/1000 bases differ from reference
>   - If reads where error free and sampling Poisson â€¦
>   - â€¦ 14x coverage would allow for 99.8% genotype accuracy
>   - â€¦ 30x coverage of the genome needed to allow for errors and clustering
> - **Population Based Prior**
>   - Uses frequency information obtained from examining other individuals
>   - Calling very rare polymorphisms still requires 20-30x coverage of the genome
>   - Calling common polymorphisms requires much less data
> 
><https://pdfs.semanticscholar.org/0156/04c3ba76cf247a8010f74ec1386e58ceb530.pdf>