# Dragon star Day 2 Pt.2

> [Dragonstar2019 by Kai Wang](https://links.jianshu.com/go?to=https%3A%2F%2Fgithub.com%2FWGLab%2Fdragonstar2019)
>
> 1. Alignment of short/long-read sequencing data
> 2. Genome assembly by short/long-read sequencing

[TOC]

## Part â…¡ Genome assembly by short/long-read sequencing

### 1 genome assembly

#### 1.1 What is genome assembly

- Genome sequence produced after chromosomes have been fragmented
- Those fragments have been sequenced
- The resulting sequences have been **put back together**

#### 1.2 Why genome assembly

- Key to understanding genetic variation
- The more accurate the reference genome, the easier it is to map reads and interpret the functional impacts of genetic mutations
- Accurate and complete genome assembly is important for human genetic 
  studies

#### 1.3 What types of genome assembly?

##### 1.3.1 *de novo* assembly

The entire assembled sequence is resolved from raw sequence data **without comparison** to a reference genome sequence.

> ![](https://media.nature.com/lw685/nature-assets/nmeth/journal/v9/n4/images/nmeth.1935-I1.jpg)
>
> Assemblers need copious sequencing data and informatic exertion to put the genome back together. *Image: Kelly Howe, Lawrence Berkeley Laboratory*

> Involves sequencing a novel genome from scratch, without the aid of external data.
>
> ![](https://s2.ax1x.com/2019/08/10/eL0mMn.png)
>
> <https://era7bioinformatics.com/en/page.cfm?id=1500>

##### 1.3.2 Comparative assembly

- You have a â€œ**reference**â€ genome to guide the assembly process
- Or sequence a **chimp genome** using the existing **human assembly** as a  guide

> Each read is aligned to the reference genome. In this approach, the new completely different sequences are lost.
>
> ![](https://s2.ax1x.com/2019/08/10/eL0cQA.png)
>
> ![](https://s2.ax1x.com/2019/08/10/eL0WeP.png)

##### 1.3.3 *De Novo* Assembly paradigms

- **Overlap-layout-consensus (OLC) methods**

  > The overlap layout consensus (OLC) method is rather intuitive and consists of a **three steps process**. 
  >
  > 1. pairwise overlaps
  > 2. layout the reads
  > 3. consensus
  >
  > ![](https://s2.ax1x.com/2019/08/10/eLOItf.png)
  >
  > A General bioinformatics background of sequence assembly. https://doi.org/10.1371/journal.pone.0169662.s001
  
  > OLC - String Graph Assemblers
  >
  > - Construct **overlap graph** directly from reads, eliminating redundant reads; trace path for assembly 
  > - Examples: SGA, Fermi
  >
  > https://ocw.mit.edu/courses/biology/7-91j-foundations-of-computational-and-systems-biology-spring-2014/lecture-slides/MIT7_91JS14_Lecture6.pdf
  
- **de Bruijn (k-mer) graph** 

  - Especially useful for assembly from short reads

  - Stacking overlapping sequences of genomic fragments of a **defined size** (the k-mer), generated by breaking each read into k-mer size.

  > de Bruijn graphs were originally proposed as a solution to the so-called "superstring problemâ€"in 1946.
  >
  > ![](https://s2.ax1x.com/2019/08/10/eLjTyQ.png)
  >
  > A General bioinformatics background of sequence assembly. https://doi.org/10.1371/journal.pone.0169662.s001

- **OLC versus de Bruijn graph**

  - OLC: pairwise alignment between reads are detected and merged
  - de Buijn: reads are decomposed into k-mers and merged
  
  ![Chaisson et al, Nature Reviews Genetics, 2015](https://s2.ax1x.com/2019/08/11/ejVze1.png)

### 2  Graph theory

- **Graph**
  
  - Mathematical structures used to model **pairwise relations** between objects
  - A graph is made up of **vertices** (also called nodes or points) which are connected by **edges** (also called links or lines).
  
  > **å›¾è®º**ï¼ˆè‹±è¯­ï¼šGraph theoryï¼‰æ˜¯[ç»„åˆæ•°å­¦](https://zh.wikipedia.org/wiki/ç»„åˆæ•°å­¦)çš„ä¸€ä¸ªåˆ†æ”¯ï¼Œå’Œå…¶ä»–æ•°å­¦åˆ†æ”¯ï¼Œå¦‚ç¾¤è®ºã€çŸ©é˜µè®ºã€æ‹“æ‰‘å­¦æœ‰ç€å¯†åˆ‡å…³ç³»ã€‚[å›¾](https://zh.wikipedia.org/wiki/å›¾_(æ•°å­¦))æ˜¯å›¾è®ºçš„ä¸»è¦ç ”ç©¶å¯¹è±¡ã€‚å›¾æ˜¯ç”±è‹¥å¹²ç»™å®šçš„**é¡¶ç‚¹**åŠè¿æ¥ä¸¤é¡¶ç‚¹çš„**è¾¹**æ‰€æ„æˆçš„å›¾å½¢ï¼Œè¿™ç§å›¾å½¢é€šå¸¸ç”¨æ¥æè¿°æŸäº›äº‹ç‰©ä¹‹é—´çš„æŸç§ç‰¹å®šå…³ç³»ã€‚é¡¶ç‚¹ç”¨äºä»£è¡¨äº‹ç‰©ï¼Œè¿æ¥ä¸¤é¡¶ç‚¹çš„è¾¹åˆ™ç”¨äºè¡¨ç¤ºä¸¤ä¸ªäº‹ç‰©é—´å…·æœ‰è¿™ç§å…³ç³»ã€‚
  >
  > https://zh.wikipedia.org/wiki/å›¾è®º
  
  > A distinction is made between **undirected graphs**, where edges link two vertices symmetrically, and **directed graphs**, where edges link two vertices asymmetrically.
  >
  > å³æ— åºå›¾å¯¹ç§°åœ°è¿æ¥ä¸¤ä¸ªè¾¹ï¼Œæœ‰åºå›¾ä¸å¯¹ç§°åœ°è¿æ¥ä¸¤ä¸ªè¾¹ã€‚
  
  ![](https://s2.ax1x.com/2019/08/10/eOOYxe.png)

#### 2.1 Directed graph

> A **directed graph** is an ordered pair *G* = (*V*, *E*) comprising:
>
> - *V* a [set](https://en.wikipedia.org/wiki/Set_(mathematics)) of *vertices* (also called *nodes* or *points*);
> - *E* âŠ† {(*x*, *y*) | (*x*, *y*) âˆˆ *V*2 âˆ§ *x* â‰  *y*} a [set](https://en.wikipedia.org/wiki/Set_(mathematics)) of *edges*
>
> <https://en.wikipedia.org/wiki/Graph_theory#Directed_graph>

- G(V, E): a set of vertices (V) and a set of directed edges (E)

- Vertex and edge may have **weights** associated with them

- Here, V= {Read1, Read2, Read3}, E = {(Read1, Read2), (Read2, Read3), 
  (Read3, Read1)} with weights of 5, 6 and 4.

  ![](https://s2.ax1x.com/2019/08/10/eXppWt.png)

#### 2.2 Overlap-based approach (OLC)

- **Overlap graph**

  - **Nodes** are initially formed by the **sequences** of the individual reads 
  - **Edges** are represented by the sequence **overlaps** between these reads.

- **Constructing the overlap graph** 

  - Create the overlap graph for the following string, where overlap means suffix/prefix match of >=5bp

    **TTATGCCATGGGATGAA**

  - Assume that we generated a few short reads from this string: TTATGCCATG, CCATGGGATG, TGGGATGAA

  - **Vertices/nodes** are formed by the sequences of the individual reads; **edges** are represented by the sequence overlaps between these reads

  ![](https://s2.ax1x.com/2019/08/10/eXkHNF.png)

  ![](https://s2.ax1x.com/2019/08/10/eXkL9J.png)

  - Sometimes graphs can be reduced by eliminating transitive edges

    ![](https://s2.ax1x.com/2019/08/10/eXVCSH.png)

    ![](https://s2.ax1x.com/2019/08/10/eXVnfg.png)

### 3 Comparison of two approaches for large genomes such as human genomes
#### 3.1 U.S. Human Genome Project

- **Primary goals** were to
  - Discover the complete set of human genes and make them accessible for further biological study
  - Determine the complete sequence of DNA bases in the human genome

- **Two thoughts**: whole-genome shotgun versus physical mapping followed by BAC-by-BAC sequencing

#### 3.2 Draft human genome

ä¸¤è‚¡åŠ›é‡ï¼š

- **Public effort**: BAC based sequencing - Performed by a consortium of government labs and universities. 
  - Break genome into ~100-300kb pieces 
  - Create Bacterial Artificial Chromosomes (BACs) from each piece
  - Assemble each piece by Sanger sequencing
  - Then stitch overlapping BACs together to draft genome
  
  ![](https://s2.ax1x.com/2019/08/11/ejSGpq.png)
- **Private effort**: whole genome random shotgun sequencing - Performed by Celera, which is a company. 
  
  - In 1998, Craig Venter announced that he was forming Celera that within three years would sequence human genome
  
  ![](https://s2.ax1x.com/2019/08/10/eXZAgJ.png)

åŒæ–¹å‘æ–‡äº’æ€¼ï¼Œç«äº‰ä¹‹ä¸‹ï¼Œ2001å¹´åˆ†åˆ«åœ¨*Nature* (2001.02.15), *Science* (2001.02.16)å‘å¸ƒæˆæœï¼Œæ—¶é—´ä»…ç›¸å·®ä¸€å¤©ï¼Œæ¯”åŸå®šè®¡åˆ’æå‰äº†ä¸¤å¹´ã€‚

#### 3.3 How Perl Saved the Human Genome Project

Random web quotes: â€œPerl and the human genome are almost perfectly matched; both are almost incomprehensible, with no central design, accreted haphazardly over a long time.â€

"Perlä¸€å¤œä¹‹é—´ç«èµ·æ¥ï¼Œäººç±»åŸºå› ç»„å®Œæˆååˆåœ¨ä¸€å¤œä¹‹é—´è¢«æŠ›å¼ƒã€‚" ğŸŒš

#### 3.4 Back to today

- Gigabase genomes can be easily sequenced and assembled using PacBio/Nanopore **long-read** assemblers
- **Whole-genome shotgun** is the predominant approach to be used today for assembling these large genomes

### 4 Lander-Waterman statistics in genome assembly
- Let 

  **N** = # of reads, 

  **L** = length of a read (average length), 

  **G** = genome length, 

  **c** = coverage = $Nâˆ—L/G$

- Main questions to address:

  - **Whatâ€™s the fraction** of genome that are covered by reads?
  - **How many contigs** are generated?
####  4.1 Coverage at a position can be modelled by Poisson distribution

$P(k, Î») = e^{-Î»} *Î»^k/k!$

- Probability that a base is **NOT covered** = $P(0,c) = e^{-c}$
- For genome size G, # of uncovered bases = $G* P(0,c) = G * e^{-c}$

#### 4.2 Whatâ€™s the fraction of genome that are covered by reads?

- ç­‰åŒäº "how many positions have coverage>0?"

- For genome size G, number of positions with read coverage:

  $G * (1-P(0,c) ) = G * (1-e^{-c})$

#### 4.3 How many contigs are generated?

- ç­‰åŒäº "how many gaps are there?", "how many reads do not have overlap > Î¸ with any other reads"

  > ![](https://s2.ax1x.com/2019/08/11/ejSmX8.png)
  >
  > <https://www.cs.cmu.edu/~ckingsf/bioinfo-lectures/sequencing.pdf>

- For each read, we calculate the probability that **zero** reads start at (1-Î¸)L

  $e^{-(1-Î¸)c} = e^{-(1-Î¸)(NL/G)}$

  å› ä¸ºï¼š$P(0,c) = e^{-c}$, $c= Nâˆ—L/G$

- Theory is different from reality !

  If there is **no repeat**, **no polymorphism**, **no region bias**, and **no sequence error**, sequence assembly could be very easy

  > ![](https://s2.ax1x.com/2019/08/11/ejp22q.png)
  >
  > <https://www.cs.cmu.edu/~ckingsf/bioinfo-lectures/sequencing.pdf>

#### 4.4 Assembling large genomes today

Since 2013, de novo assembly of large genomes has shifted from **short-read** sequencing to **synthetic** or true **long-read** sequencing.

![Jung et al, Trends in Plant Science, 2019](https://s2.ax1x.com/2019/08/11/ejpLxx.png)

#### 4.5 Canu assembler for long-read assembly

åŸºäºCelaraå¼€å‘çš„è½¯ä»¶

![Koren et al, Genome Research, 2017](https://s2.ax1x.com/2019/08/11/ejpzZD.png)

#### 4.6 Wtdbg2 assembler for long-read assembly

Wtdbg2 groups 256 base pairs into a bin. (å¹¶ä¸æ˜¯ç›´æ¥ç”¨ATCG)

![Ruan et al, BioRxiv, 2019](https://s2.ax1x.com/2019/08/11/ej9Sde.png)

### 5 Seven Bridges of KÃ¶nigsberg Problem

> **æŸ¯å°¼æ–¯å ¡ä¸ƒæ¡¥é—®é¢˜**ï¼ˆSeven Bridges of KÃ¶nigsbergï¼‰æ˜¯[å›¾è®º](https://zh.wikipedia.org/wiki/å›¾è®º)ä¸­çš„è‘—åé—®é¢˜ã€‚è¿™ä¸ªé—®é¢˜æ˜¯åŸºäºä¸€ä¸ªç°å®ç”Ÿæ´»ä¸­çš„äº‹ä¾‹ï¼šå½“æ—¶[ä¸œæ™®é²å£«](https://zh.wikipedia.org/wiki/æ±æ™®é­¯å£«)[æŸ¯å°¼æ–¯å ¡](https://zh.wikipedia.org/wiki/æŸ¯å°¼æ–¯å ¡)ï¼ˆä»Šæ—¥[ä¿„ç½—æ–¯](https://zh.wikipedia.org/wiki/ä¿„ç¾…æ–¯)[åŠ é‡Œå®æ ¼å‹’](https://zh.wikipedia.org/wiki/åŠ é‡Œå¯§æ ¼å‹’)ï¼‰å¸‚åŒºè·¨[æ™®åˆ—æˆˆåˆ©äºšæ²³](https://zh.wikipedia.org/wiki/æ™®åˆ—æˆˆåˆ©äºšæ²³)ä¸¤å²¸ï¼Œæ²³ä¸­å¿ƒæœ‰ä¸¤ä¸ªå°å²›ã€‚å°å²›ä¸æ²³çš„ä¸¤å²¸æœ‰ä¸ƒæ¡æ¡¥è¿æ¥ã€‚åœ¨æ‰€æœ‰æ¡¥éƒ½åªèƒ½èµ°ä¸€éçš„å‰æä¸‹ï¼Œå¦‚ä½•æ‰èƒ½æŠŠè¿™ä¸ªåœ°æ–¹æ‰€æœ‰çš„æ¡¥éƒ½èµ°éï¼Ÿ
>
> https://zh.wikipedia.org/wiki/æŸ¯å°¼æ–¯å ¡ä¸ƒæ¡¥é—®é¢˜

![](https://s2.ax1x.com/2019/08/11/ej9twF.png)

- In 1736, the Swiss mathematician **Leonhard Euler** demonstrated that the answer is **no** å°±æ˜¯é«˜æ•°é‡Œæ¬§æ‹‰å…¬å¼çš„é‚£ä¸ªæ¬§æ‹‰

- In graph theory

  - An **Eulerian trail** (or Eulerian path) is a trail in a finite graph which visits every edge exactly once (allowing for revisiting vertices).

    > [è±æ˜‚å“ˆå¾·Â·æ¬§æ‹‰](https://zh.wikipedia.org/wiki/è±æ˜‚å“ˆå¾·Â·æ¬§æ‹‰)åœ¨1735å¹´æå‡ºï¼Œå¹¶æ²¡æœ‰æ–¹æ³•èƒ½åœ†æ»¡è§£å†³è¿™ä¸ªé—®é¢˜ï¼Œä»–æ›´åœ¨ç¬¬äºŒå¹´å‘è¡¨åœ¨[è®ºæ–‡](https://zh.wikipedia.org/wiki/è®ºæ–‡)ã€ŠæŸ¯å°¼æ–¯å ¡çš„ä¸ƒæ¡¥ã€‹ä¸­ï¼Œè¯æ˜ç¬¦åˆæ¡ä»¶çš„èµ°æ³•å¹¶ä¸å­˜åœ¨ï¼Œä¹Ÿé¡ºå¸¦æå‡ºå’Œè§£å†³äº†[ä¸€ç¬”ç”»é—®é¢˜](https://zh.wikipedia.org/wiki/ä¸€ç¬”ç”»é—®é¢˜)ã€‚
    >
    > https://zh.wikipedia.org/wiki/æŸ¯å°¼æ–¯å ¡ä¸ƒæ¡¥é—®é¢˜

  - Similarly, an **Eulerian circuit** or Eulerian cycle is an Eulerian trail which starts and ends on the **same vertex**. 

    ![](https://s2.ax1x.com/2019/08/11/ej9RFH.png)

#### 5.1 Eulerâ€™s Theorem on directed graphs

- For **directed graphs**, the cycle will need to follow the direction of the edges

  - Indegree(v) = # edges coming into v
  - Outdegree(v) = # edges leaving v

- A directed graph has an **Eulerian path** if and only if: 
  **indegree(v)=outdegree(v)** for all but 2 nodes (x and y), where indegree(x)=outdegree(x)+1, and indegree(y)=outdegree(y)-1. 

  è‡³å¤šæœ‰ä¸€ä¸ªèŠ‚ç‚¹ **è¿›>å‡º**ï¼Œè‡³å¤šæœ‰ä¸€ä¸ªèŠ‚ç‚¹ **å‡º>è¿›**

- A directed graph has an **Eulerian cycle** if and only if **indegree(v)=outdegree(v) for all nodes**

  æ‰€æœ‰èŠ‚ç‚¹çš„degreeæ•°å„è‡ªç›¸ç­‰

#### 5.2 Eulerâ€™s Theorem on undirected graphs

- For undirected graphs, the degree of a vertex is the number of edges that are incident to the vertex 

- An undirected graph has an **Eulerian trail** if and only if exactly zero or two vertices have odd degree

  å¿…é¡»æœ‰0ä¸ªæˆ–2ä¸ªèŠ‚ç‚¹çš„degreeä¸ºå¥‡æ•°

- An undirected graph has an **Eulerian cycle** if and only if every vertex has **even degree** 

  æ‰€æœ‰èŠ‚ç‚¹çš„degreeä¸ºå¶æ•°

#### 5.3 The KÃ¶nigsberg graph is not Eulerian

For the existence of Eulerian trails 

- It is necessary that zero or two vertices have an odd degree âŒ ABCDçš„degreeéƒ½ä¸ºå¥‡æ•°

- If there are no vertices of odd degree, all Eulerian trails are circuits âŒ
- If there are exactly two vertices of odd degree, all Eulerian trails start at one of them and end at the other âŒ

![](https://s2.ax1x.com/2019/08/11/ejktPg.png)

æ‰€ä»¥æ¬§æ‹‰è®¤ä¸ºä¸ƒæ¡¥é—®é¢˜æ— è§£

#### 5.4 Hamiltonian vs Eulerian path

- **Hamiltonian path**: a path that travels to every node exactly **once** and **ends at the starting node**
- Eulerian path: a path that visits all edges of a graph exactly **once**
- When analyzing sequencing data, instead of assigning each k-mer contained in some read to a node, we will assign each such **k-mer to an edge**. This allows the construction of a '**de Bruijn graph**'
- The **vertices are (k-1)-mers** that appear in some read, and **edges defined by overlap of k-2 nucleotides**

### 6 Constructing the de Bruijn Graph 

Create the de Bruijn graph for the following string, using **k=3**

**TTATGCCATGGGATGAA**

![](https://s2.ax1x.com/2019/08/11/ejANTK.png)

![](https://s2.ax1x.com/2019/08/11/ejAXhF.png)

#### 6.1 Error correction in de Bruijn graph

- Erroneous reads (marked in red) introduce *k* false *k*-mers to the graph, resulting in additional spurious branches, so-called "bubbles" and "tips" 

  ![](https://s2.ax1x.com/2019/08/11/ejZGOs.png)

- Influence of *k*-mer length on assembly

  *k*å¤ªå¤§å¯¼è‡´ä¸¤æ¡è·¯å¾„æ²¡æœ‰é‡å  or *k*å¤ªå°å¯¼è‡´å¯èƒ½çš„ç»“æœå¤ªå¤š

  ![](https://s2.ax1x.com/2019/08/11/ejZBpF.png)

- What is the appropriate k-mer?

  - Tools such as [kmerGenie](http://kmergenie.bx.psu.edu/) and [Velvet Advisor](dna.med.monash.edu.au/~torsten/velvet_advisor) have been developed to predict the optimal value for k for a given read dataset, based on read length and k-mer frequencies
  - Other strategies may include the **merging of assemblies** produced with different k-mer sizes or the **use of a multi k-mer** assembly method

#### 6.2 A fun hypothetical case study: Frederick Sangerâ€™s insulin sequencing study
"Sangeræ˜¯ä¸ªå¾ˆæœ‰è¶£çš„äººï¼Œæ²¡æœ‰å­¦æ ¡è¦ä»–ï¼Œä½†æ˜¯ä»–å®¶é‡Œå¾ˆæœ‰é’±ï¼Œè‡ªå·±è¿›å®éªŒå®¤è‡ªå·±æé’±ä¹°è¯•å‰‚ï¼Œå¾—äº†è¯ºè´å°”å¥–åè¿™è¾ˆå­ä¹Ÿå°±è¿™ä¹ˆå›äº‹äº†ï¼Œä¹Ÿä¸å‘æ–‡ç« ä¹Ÿä¸çœ‹æ–‡çŒ®...."

![*ç©·é¬¼çˆ†ç¬‘ ](https://wx3.sinaimg.cn/mw690/695f6845ly1g5hpyy8xrvj204e05imxa.jpg)

![](https://s2.ax1x.com/2019/08/11/ejmFrd.png)

ç”¨ de Bruijn graph çš„æ–¹æ³•ï¼Œä¹Ÿå¯ä»¥å°†äºŒè‚½åŠä¸‰è‚½ç‰‡æ®µç»„è£…å‡ºèƒ°å²›ç´ contig.